---
title: "Methyl_paper_1_2_2018"
author: "Swapna Mahurkar-Joshi"
date: "January 2, 2018"
output: html_document
---

```{r setup}
# Set working directory outside the chunks
knitr::opts_knit$set(root.dir = normalizePath("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/"))
```

# DMPs in IBS

```{r}
load("data1/consistant_data/colon_meth_level4.rda")
#load("data1/consistant_data/GRsetPbmcColFunFlt.rda")
library(knitr)
library(DMRcate)
```

```{r}
ann450kSub  <- col_meth_level4[["ann450kSub"]] 
metDat_c    <- col_meth_level4[["metDat_c"]]; dim(metDat_c)
colon_betas <- col_meth_level4[["col_betas"]]
rm(col_meth_level4)

metDat_c$BH <-  gsub("U", "M", metDat_c$BH)
ibs    <- subset(metDat_c, metDat_c$Group==2)
hc     <- subset(metDat_c, metDat_c$Group==1)
```

# colon
##Table 1: Demographics for colon data

```{r}
table1 <- matrix(NA, nrow = 16, ncol = 3)

table1[c(1,2,5:16), 1] <- round(apply(ibs[ ,c(4,5,6,7,8,9,14,15,16, 17,18,21,25,27)], 2, mean, na.rm = TRUE), 2)
table1[c(1,2,5:16), 2] <- round(apply(hc[ ,c(4,5,6,7,8,9,14,15,16,17,18,21,25,27)], 2, mean, na.rm = TRUE), 2)
table1[3,1] <- round(table(ibs$Sex)[2]/(table(ibs$Sex)[2]+table(ibs$Sex)[1])*100,2)
table1[3,2] <- round(table(hc$Sex)[2]/(table(hc$Sex)[2]+table(hc$Sex)[1])*100,2)

row.names(table1)[c(1,2,5:16)] <- colnames(hc[, c(4,5,6,7,8,9,14,15,16,17,18,21,25,27)])
row.names(table1)[3]<-"Female %"
row.names(table1)[4]<-"Bowel Habit %"

table1[c(1,2,5:16), 1] <- round(apply(ibs[ ,c(4,5,6,7,8,9,14,15,16, 17,18,21,25,27)], 2, sd, na.rm = TRUE), 2)
table1[c(1,2,5:16), 2] <- round(apply(hc[ ,c(4,5,6,7,8,9,14,15,16, 17,18,21,25,27)], 2, sd, na.rm = TRUE), 2)
```

```{r}
metDat_c$Group <- as.factor(metDat_c$Group)
comb1          <- expand.grid(names(metDat_c[,c(4,5,14,15,16,17,18,25,27)]),"Group")

models     <- lapply(paste(comb1[,1],comb1[,2], sep = "~"), formula)
res.models <- lapply(models, FUN = function(x) {t.test(formula = x, data = metDat_c)})
# names(res.models) <- paste(comb1[,1],comb1[,2], sep = "~")
table1[c(1:2,5:11),3] <- round(t(data.frame(lapply(res.models, FUN = t_p))),5)
colnames(table1) <- c("IBS (N=102)","Healthy Controls (N=36)", "t.test p-value")
table2 <- matrix(as.numeric(as.character(table1)), nrow = 11, ncol = 3)
row.names(table2) <- row.names(table1)
colnames(table2)  <- colnames(table1)
table2 <-  as.data.frame(table2)
table2[c(5,6,9,14),c(2,3)] <- "-"
# Fisher test for sex and bowel habits
fisher.test(metDat_c$Group, metDat_c$Sex)
table2[3,3] <- 0.318
ibsBHper    <- c(round(table(ibs$BH)[1]/dim(ibs)[1]*100,2),round(table(ibs$BH)[2]/dim(ibs)[1]*100,2),round(table(ibs$BH)[3]/dim(ibs)[1]*100,2),round(table(ibs$BH)[4]/dim(ibs)[1]*100,2))
table2[3,1] <- paste(paste("IBS-C",ibsBHper[1],sep = "="),paste("IBS-D",ibsBHper[2],sep = "="),paste("IBS-M",ibsBHper[3],sep = "="), sep=",")

kable(table2,align=c(rep('c', 14)), caption = "Table1: Clinical Characteristics of IBS Patients and HCs")
```

```{r}
colDxDMPs <- dmpFinder(getM(GRsetColFunFlt),pData(GRsetColFunFlt)$Dx, type = "categorical", qCutoff = 1, shrinkVar = TRUE)
betas_col <- getBeta(GRsetColFunFlt)
md1 <- rowMeans(betas_col[,colnames(betas_col)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$Dx ==  levels(as.factor(pData(GRsetColFunFlt)$Dx))[2]))]) - rowMeans(betas_col[,colnames
(betas_col)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$Dx ==  levels(as.factor(pData(GRsetColFunFlt)$Dx))[1]))])
colDxDMPs$meanDiff <- md1
```

```{r}
colDxDMPs   <- colDxDMPs[row.names(col_meth_level4[["ann450kSub"]]),]
if(all.equal(row.names(colDxDMPs), row.names(col_meth_level4[["ann450kSub"]]))== TRUE){
colDxDMPsAn <- cbind(colDxDMPs, col_meth_level4[["ann450kSub"]])
}
colDxDMPsAn <- colDxDMPsAn[order(colDxDMPsAn$pval),]
```

```{r}
colDxDMPs_sig      <- subset(colDxDMPsAn, colDxDMPsAn$qval<0.05); dim(colDxDMPs_sig)
colDxDMPs_sig_p05  <- subset(colDxDMPsAn, colDxDMPsAn$pval<0.05); dim(colDxDMPs_sig_p05)
colDxDMPs_sig_p001 <- subset(colDxDMPsAn, colDxDMPsAn$pval<0.001); dim(colDxDMPs_sig_p001)
# Suggestive DMPs (p<5e-05)
colDxDMPs_sig_sug  <- subset(colDxDMPsAn, colDxDMPsAn$pval<0.00005); dim(colDxDMPs_sig_sug)
```

# Plot top 4 differntially methylated CpGs 

```{r}
cpgs <- rownames(colDxDMPs_sig_p001)[1:4]
par(mfrow=c(2,2))
plotCpg(getBeta(GRsetColFunFlt), cpg=cpgs, pheno=pData(GRsetColFunFlt)$Dx)
```

```{r}
save(colDxDMPsAn, file= "temp1/res3_functional_norm/methBioconductorWorkflow/colon/ColDxDMPsAn.rda")
write.csv(colDxDMPs_sig_p001, file= "temp1/res3_functional_norm/methBioconductorWorkflow/colon/colDxDMPs_sig_p001_231.csv")
write.csv(colDxDMPs_sig_sug, file= "temp1/res3_functional_norm/methBioconductorWorkflow/colon/colDxDMPs_sig_sug.csv")
```


```{r}
library(missMethyl)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)

gst.KEGG <- gometh(sig.cpg=rownames(colDxDMPs_sig_p001), all.cpg=rownames(colDxDMPsAn), collection="KEGG", prior.prob = T)
gst.KEGG <- gst.KEGG[order(gst.KEGG$P.DE),]
gst.KEGG  <- gst.KEGG[gst.KEGG$FDR<0.05,]

ann <- missMethyl:::.flattenAnn(fit2)
m <- match(rownames(fit2), ann$cpg)
EntrezID <- ann$entrezid[m]
n <- table(EntrezID)
k <- kegga(fit2, geneid=EntrezID, covariate=n[EntrezID])
topKEGG(k)
```

# DMRs

```{r}
library(DMRcate)
library(minfi)
library(limma)
targets <- pData(GRsetColFunFlt)
dx <- factor(targets$Dx)
design <- model.matrix(~0+dx, data=targets)
colnames(design) <- levels(dx)
contMatrix <- makeContrasts(IBS-HC,   levels=design)
contMatrix

myAnnotation <- cpg.annotate(object = getM(GRsetColFunFlt), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "IBS - HC", arraytype = "450K")

str(myAnnotation)

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]
```

# convert the regions to annotated genomic ranges

```{r}
results.ranges <- extractRanges(DMRs, genome = "hg19")
save(results.ranges, file = "temp1/res3_functional_norm/methBioconductorWorkflow/colon/colDMRs_p001.rda")
write.csv(results.ranges, file = "temp1/res3_functional_norm/methBioconductorWorkflow/colon/colDMRs_p001.csv")
```

# set up the grouping variables and colours and plot

```{r}
library(RColorBrewer)
data(dmrcatedata)
pal <- brewer.pal(3,"Dark2")
groups <- pal[1:length(unique(targets$Dx))]
names(groups) <- levels(factor(targets$Dx))
cols <- groups[as.character(factor(targets$Dx))]
samps <- 1:nrow(targets)

# draw the plot for the top DMR
png("temp1/res3_functional_norm/methBioconductorWorkflow/colon/ColDxDMR2.png", height = 2000, width = 2000, res = 250)
DMR.plot(ranges=results.ranges, dmr=2, CpGs=getBeta(GRsetColFunFlt), phen.col=cols, what = "Beta",      arraytype = "450K", pch=16, toscale=FALSE, plotmedians=TRUE,   genome="hg19", samps=samps)
dev.off()
```

# plotting DMRs with Gviz
## setup tracks

```{r}
library(Gviz)
load("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/Colon/ColDxDMPs.rda")
## STC2 DMR
STC2SigCpgs <- ColDxDMPs_sig_p05[ColDxDMPs_sig_p05$UCSC_RefGene_Name =="STC2",]
STC2SigCpgsPlusStr <- subset(STC2SigCpgs, STC2SigCpgs$strand == "+")
STC2SigCpgsPlusStr <- STC2SigCpgsPlusStr [-4,]
STC2SigCpgsNegStr <- subset(STC2SigCpgs, STC2SigCpgs$strand == "-")

STC2SigCpgs <- STC2SigCpgsPlusStr
STC2SigCpgs <- STC2SigCpgs[order(STC2SigCpgs$pos),]
betas_STC2dmr <- getBeta(GRsetPbmcFunFlt)[row.names(getBeta(GRsetPbmcFunFlt))%in%row.names(STC2SigCpgs),]
dx <- as.factor(pData(GRsetPbmcFunFlt)$Dx)
betas_STC2dmr <- betas_STC2dmr[row.names(STC2SigCpgs),]
grObj <- makeGRangesFromDataFrame(data.frame(chr = STC2SigCpgs$chr, start = STC2SigCpgs$pos, end = STC2SigCpgs$pos+1, strand = STC2SigCpgs$strand, score = STC2SigCpgs$meanDiff))

atrack <- AnnotationTrack(grObj, name = "CpG")
plotTracks(atrack)
chromosome(atrack) <- 'chr5'

gtrack <- GenomeAxisTrack()
plotTracks(list(gtrack, atrack))

# genome : "hg19" 
# gen<-genome(grObj)

gen = "hg19" 
chr <- as.character(unique(seqnames(grObj)))

# Ideogram track
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
plotTracks(list(itrack, gtrack, atrack))
# geneModels <- data.frame(c(rep("chr20",2)), c(58206655,58205944),c(58303445,58299695),c(96790,93751), c("+","+"),c("protein coding","Nonsense mediated decay"),c("ENSG00000118946" , "ENSG00000118946"),c("ENSG00000118946" , "ENSG00000118946") ,c("ENST00000377918","ENST00000484979"),c("PCDH17","PCDH17"))
# colnames(geneModels) <- c( "chromosome",    "start",      "end", "width", "strand", "feature","gene",  "exon",      "transcript",     "symbol")

# grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr, name = "Gene Model")
# chromosome(grtrack) <- 'chr13'
# plotTracks(list(itrack, gtrack, atrack, grtrack))

##data track
 # coords <- c(rep(58205001,145),rep(58205045,145), rep(58209133,145))
mcols(grObj) <- betas_STC2dmr
 dxCol <- ifelse(dx=="HC","red","green")
dtrack <- DataTrack(grObj,groups = as.vector(dx),type= c("b"),aggregateGroups = TRUE,
           aggregation = "mean")
# 
# plotTracks(dtrack, groups = as.vector(bh), type= "b", aggregateGroups = TRUE,
#            aggregation = "max")
plotTracks(list(itrack, gtrack, atrack, dtrack))
```

```{r}
SLC38A4SigCpgs <- ColDxDMPs_sig_p05[grep("SLC38A4",ColDxDMPs_sig_p05$UCSC_RefGene_Name),]; dim(SLC38A4SigCpgs)
SLC38A4SigCpgsPlusStr <- subset(SLC38A4SigCpgs, SLC38A4SigCpgs$strand == "+")
# SLC38A4SigCpgsPlusStr <- SLC38A4SigCpgsPlusStr [-4,]
SLC38A4SigCpgsNegStr <- subset(SLC38A4SigCpgs, SLC38A4SigCpgs$strand == "-")

SLC38A4SigCpgs <- SLC38A4SigCpgsPlusStr
SLC38A4SigCpgs <- SLC38A4SigCpgs[order(SLC38A4SigCpgs$pos),]
betas_SLC38A4dmr <- getBeta(GRsetPbmcFunFlt)[row.names(getBeta(GRsetPbmcFunFlt))%in%row.names(SLC38A4SigCpgs),]
dx <- as.factor(pData(GRsetPbmcFunFlt)$Dx)
betas_SLC38A4dmr <- betas_SLC38A4dmr[row.names(SLC38A4SigCpgs),]
grObj <- makeGRangesFromDataFrame(data.frame(chr = SLC38A4SigCpgs$chr, start = SLC38A4SigCpgs$pos, end = SLC38A4SigCpgs$pos+1, strand = SLC38A4SigCpgs$strand, score = SLC38A4SigCpgs$meanDiff))


atrack <- AnnotationTrack(grObj, name = "CpG")
plotTracks(atrack)
chromosome(atrack) <- 'chr12'

gtrack <- GenomeAxisTrack()
plotTracks(list(gtrack, atrack))

# genome : "hg19" 
# gen<-genome(grObj)

gen = "hg19" 
chr <- as.character(unique(seqnames(grObj)))

# Ideogram track
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
plotTracks(list(itrack, gtrack, atrack))
# geneModels <- data.frame(c(rep("chr20",2)), c(58206655,58205944),c(58303445,58299695),c(96790,93751), c("+","+"),c("protein coding","Nonsense mediated decay"),c("ENSG00000118946" , "ENSG00000118946"),c("ENSG00000118946" , "ENSG00000118946") ,c("ENST00000377918","ENST00000484979"),c("PCDH17","PCDH17"))
# colnames(geneModels) <- c( "chromosome",    "start",      "end", "width", "strand", "feature","gene",  "exon",      "transcript",     "symbol")

# grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr, name = "Gene Model")
# chromosome(grtrack) <- 'chr13'
# plotTracks(list(itrack, gtrack, atrack, grtrack))

##data track
 # coords <- c(rep(58205001,145),rep(58205045,145), rep(58209133,145))
mcols(grObj) <- betas_SLC38A4dmr
 dxCol <- ifelse(dx=="HC","red","green")
dtrack <- DataTrack(grObj,groups = as.vector(dx),type= c("b"),aggregateGroups = TRUE,
           aggregation = "mean")
# 
# plotTracks(dtrack, groups = as.vector(bh), type= "b", aggregateGroups = TRUE,
#            aggregation = "max")
plotTracks(list(itrack, gtrack, atrack, dtrack))
```

########################################################################################
# BH associated DMPs Colon
########################################################################################

```{r}
load("data1/consistant_data/GRsetPbmcColFunFlt.Rda")
load("data1/consistant_data/hm450annotations.rda")
ann450kSub <- hm450annotations[match(featureNames(GRsetColFunFlt),hm450annotations$Name),c(1:4,12:19,24:ncol(hm450annotations))] 

mVals_c <- getM(GRsetColFunFlt)
head(mVals_c[,1:5])
betaVals_c <- getBeta(GRsetColFunFlt)
head(betaVals_c[,1:5])

# factor of interest
bh <- factor(pData(GRsetColFunFlt)$BH)
```


```{r}
GRsetColFunFltCD <- GRsetColFunFlt[,colnames(GRsetColFunFlt)%in%c(row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH == "C"),]), row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH=="D"),]))]

library(minfiData)
if (require(minfiData)) {

grp <- pData(GRsetColFunFltCD)$BH
M   <- getM(GRsetColFunFltCD)
B   <- getBeta(GRsetColFunFltCD)
ColBhDMPs <- dmpFinder(M, pheno=grp, type="categorical", shrinkVar = TRUE)
ColBhDMPs$Bdiff <- rowMeans(B[row.names(ColBhDMPs),colnames(B)%in%row.names(subset(pData(GRsetColFunFltCD), pData(GRsetColFunFltCD)$Dx == "C"))])-rowMeans(B[row.names(ColBhDMPs),colnames(B)%in%row.names(subset(pData(GRsetColFunFltCD), pData(GRsetColFunFltCD)$BH == "D"))])
sum(ColBhDMPs$qval < 0.05, na.rm=TRUE)
head(ColBhDMPs)
ann450kSub <- ann450kSub[row.names(ann450kSub)%in%row.names(ColBhDMPs),]
ann450kSub <- ann450kSub[row.names(ColBhDMPs),]
ColBhDMPs <- cbind(ColBhDMPs,ann450kSub)
}

```



```{r}

# variable to correct for- could potenitally correct for age, but since there is no significant differnces in the age between ibs and hcs, we will not include it as  a covariate.

# use the above to create a design matrix
design <- model.matrix(~ 0 + bh, data=pData(GRsetColFunFlt))
colnames(design) <- c("bhC", "bhD", "bhM", "bhN")

# fit the linear model 
library(limma)
fit <- lmFit(mVals_c, design)

# create a contrast matrix for specific comparisons
contMatrix <- makeContrasts(bhC-bhN, bhD-bhN,bhM-bhN, bhC-bhD,bhM-bhD, bhM-bhC,levels=design)
contMatrix

# fit the contrasts
fit2 <- contrasts.fit(fit, contMatrix)
fit2 <- eBayes(fit2,trend=TRUE)

# look at the numbers of DM CpGs at FDR < 0.05
summary(decideTests(fit2))
#  bhC - bhN bhD - bhN bhM - bhN bhC - bhD bhM - bhD bhM - bhC
# -1         0         0         0         0         0         0
# 0     420257    420257    420257    420257    420257    420257
# 1          0         0         0         0         0         0

# get the table of results for the first contrast 
DMPs_CN <-  topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)
DMPs_CN$meanDiff <- (rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$BH ==  "C"))]))-(rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$BH ==  "N"))]))
DMPs_CNp001 <- subset(DMPs_CN, DMPs_CN$P.Value<0.001); dim(DMPs_CNp001)
# [1] 297  33

DMPs_DN <-  topTable(fit2, num=Inf, coef=2, genelist=ann450kSub)
DMPs_DN$meanDiff <- (rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$BH ==  "D"))]))-(rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$BH ==  "N"))]))
DMPs_DNp001 <- subset(DMPs_DN, DMPs_DN$P.Value<0.001); dim(DMPs_DNp001)
# [1] 139  33

DMPs_MN <-  topTable(fit2, num=Inf, coef=3, genelist=ann450kSub)
DMPs_MN$meanDiff <- (rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$BH ==  "M"))]))-(rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$BH ==  "N"))]))
DMPs_MNp001 <- subset(DMPs_MN, DMPs_MN$P.Value<0.001); dim(DMPs_MNp001)
# [1] 213  33

DMPs_CD <- topTable(fit2, num=Inf, coef=4, genelist = ann450kSub)
DMPs_CD$meanDiff <- (rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$BH ==  "C"))]))-(rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$BH ==  "D"))]))
DMPs_CD_sig <- subset(DMPs_CD, DMPs_CD$adj.P.Val <0.05); dim(DMPs_CD_sig)
DMPs_CDp001 <- subset(DMPs_CD, DMPs_CD$P.Value<0.001); dim(DMPs_CDp001)
# [1] 115  32
# DMPs_CDp05 <- subset(DMPs_CD, DMPs_CD$P.Value<0.05); dim(DMPs_CDp05)

GRsetColFunFltCD <- GRsetColFunFlt[,colnames(GRsetColFunFlt)%in%c(row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH == "C"),]), row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH=="D"),]))]


if (require(minfiData)) {
grp <- pData(GRsetPbmcFunFltCD)$BH
M <- getM(GRsetPbmcFunFltCD)
dmpCD <- dmpFinder(M, pheno=grp, type="categorical", shrinkVar = TRUE)
sum(dmpCD$qval < 0.05, na.rm=TRUE)
head(dmpCD)

}
setwd("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/BH/")
write.csv(DMPs_CNp001, file = "DMPs_BH_CNp001_col.csv")
write.csv(DMPs_DNp001, file = "DMPs_BH_DNp001_col.csv")
write.csv(DMPs_MNp001, file = "DMPs_BH_MNp001_col.csv")
write.csv(DMPs_CDp001, file = "DMPs_BH_CDp001_col.csv")

save(DMPs_CN,DMPs_DN,DMPs_MN,DMPs_CD, file = "DMPs_BH_col.Rda")
```

# DMRs

# DMRs # C and D only
```{r}

# targets <- pData(GRsetPbmcFunFltCD)
bh <- factor(pData(GRsetColFunFltCD)$BH)
design <- model.matrix(~0+bh, data=pData(GRsetColFunFltCD))
colnames(design) <- levels(bh)
contMatrix <- makeContrasts(C-D, levels=design)
contMatrix
myAnnotation <- cpg.annotate(object = getM(GRsetColFunFltCD), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "C - D", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : Factor w/ 420257 levels "cg00000029","cg00000108",..: 222425 374552 248925 18571 277135 194037 362374 266425 100898 55258 ...
#  $ stat  : num [1:420257] 0.988 0.626 -0.511 -0.166 0.138 ...
#  $ CHR   : Factor w/ 22 levels "chr1","chr10",..: 1 1 1 1 1 1 1 1 1 1 ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 0.00969 0.004969 -0.003749 -0.000111 0.000572 ...
#  $ indfdr: num [1:420257] 0.99 0.991 0.991 0.997 0.998 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
results.ranges.sort <- data.frame(results.ranges)[order(-data.frame(results.ranges)$no.cpgs),]
write.csv(results.ranges.sort, file= "temp1/res3_functional_norm/methBioconductorWorkflow/colon/BH/DMR_bowelHabits_CD_col.csv")

```

# DMRs # C and N only
```{r}
GRsetColFunFltCN <- GRsetColFunFlt[,colnames(GRsetColFunFlt)%in%c(row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH == "C"),]), row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH=="N"),]))]
# targets <- pData(GRsetPbmcFunFltCN)
bh <- factor(pData(GRsetColFunFltCN)$BH)
design <- model.matrix(~0+bh, data=pData(GRsetColFunFltCN))
colnames(design) <- levels(bh)
contMatrix <- makeContrasts(C-N, levels=design)
contMatrix
myAnnotation <- cpg.annotate(object = getM(GRsetColFunFltCN), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "C - N", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : chr [1:420257] "cg13869341" "cg24669183" "cg15560884" "cg01014490" ...
#  $ stat  : num [1:420257] 1.4234 -2.3471 -0.9021 0.0985 -0.2728 ...
#  $ CHR   : chr [1:420257] "chr1" "chr1" "chr1" "chr1" ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 1.09e-02 -1.73e-02 -6.67e-03 9.47e-05 -1.29e-04 ...
#  $ indfdr: num [1:420257] 1 1 1 1 1 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
results.ranges.sort <- data.frame(results.ranges)[order(-data.frame(results.ranges)$no.cpgs),]
write.csv(results.ranges.sort, file= "temp1/res3_functional_norm/methBioconductorWorkflow/colon/BH/DMR_bowelHabits_CN_Col.csv")
```
# DMRs # D and N only
```{r}
GRsetColFunFltDN <- GRsetColFunFlt[,colnames(GRsetColFunFlt)%in%c(row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH == "D"),]), row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH=="N"),]))]
# targets <- pData(GRsetColFunFltDN)
bh <- factor(pData(GRsetColFunFltDN)$BH)
design <- model.matrix(~0+bh, data=pData(GRsetColFunFltDN))
colnames(design) <- levels(bh)
contMatrix <- makeContrasts(D-N, levels=design)
contMatrix
myAnnotation <- cpg.annotate(object = getM(GRsetColFunFltDN), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "D - N", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : chr [1:420257] "cg13869341" "cg24669183" "cg15560884" "cg01014490" ...
#  $ stat  : num [1:420257] 0.405 -0.749 -0.424 1.054 1.149 ...
#  $ CHR   : chr [1:420257] "chr1" "chr1" "chr1" "chr1" ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 0.003585 -0.005674 -0.002822 0.000616 0.004502 ...
#  $ indfdr: num [1:420257] 1 1 1 1 1 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
results.ranges.sort<- data.frame(results.ranges)[order(-data.frame(results.ranges)$no.cpgs),]
write.csv(results.ranges.sort, file = "C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/BH/DMR_bowelHabits_DN_col.csv")
```
# DMRs # M and N only
```{r}
GRsetColFunFltMN <- GRsetColFunFlt[,colnames(GRsetColFunFlt)%in%c(row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH == "M"),]), row.names(pData(GRsetColFunFlt)[which(pData(GRsetColFunFlt)$BH=="N"),]))]
# targets <- pData(GRsetColFunFltMN)
bh <- factor(pData(GRsetColFunFltMN)$BH)
design <- model.matrix(~0+bh, data=pData(GRsetColFunFltMN))
colnames(design) <- levels(bh)
contMatrix <- makeContrasts(M-N, levels=design)
contMatrix
myAnnotation <- cpg.annotate(object = getM(GRsetColFunFltMN), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "M - N", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : chr [1:420257] "cg13869341" "cg24669183" "cg15560884" "cg01014490" ...
#  $ stat  : num [1:420257] 0.984 -0.473 -1.895 -0.734 0.205 ...
#  $ CHR   : chr [1:420257] "chr1" "chr1" "chr1" "chr1" ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 7.91e-03 -5.15e-03 -1.37e-02 5.13e-05 1.67e-03 ...
#  $ indfdr: num [1:420257] 0.965 0.978 0.965 0.967 0.991 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
results.ranges.sort<- data.frame(results.ranges)[order(-data.frame(results.ranges)$no.cpgs),]
write.csv(results.ranges.sort, file= "temp1/res3_functional_norm/methBioconductorWorkflow/colon/BH/DMR_bowelHabits_MN_col.csv")
```

# Heatmap and boxplot

```{r}
# heatmap
png("temp/functional_norm/methBioconductorWorkflow/colon/ColDxp001heatmap.png", height = 2000, width = 2000, res = 250)
heatmapMeth(GRsetColFunFlt, row.names(ColDxDMPs_sig_p001), pData(GRsetColFunFlt)$Dx, pData(GRsetColFunFlt)$BH, pData(GRsetColFunFlt)$Gender)
dev.off()

# box plots
library(ggplot2)
# plot the top 4 most significantly differentially methylated CpGs 
ColDxDMPs_sig_p001 <- ColDxDMPs_sig_p001[order(-ColDxDMPs_sig_p001$meanDiff),]
dmp <- 1:16
df <- cbind(as.data.frame(t(getBeta(GRsetColFunFlt)[row.names(GRsetColFunFlt)%in%rownames(ColDxDMPs_sig_p001)[dmp],])), pData(GRsetColFunFlt)$BH, pData(GRsetColFunFlt)$Dx) 

colnames(df)[-c(dim(df)[2],dim(df)[2]-1)] <- ColDxDMPs_sig_p001[dmp,][,26]
colnames(df)[which(duplicated(colnames(df)))] <- paste(colnames(df)[which(duplicated(colnames(df)))],"_1",sep="")
colnames(df)[dim(df)[2]] <- "Dx"
colnames(df)[dim(df)[2]-1] <- "BH"
colnames(df) <- gsub("-","_",colnames(df))

library(gridExtra)   
png("temp/functional_norm/methBioconductorWorkflow/colon/Col_Dx001p_dmp1_16.png", height = 3000, width  =3000,res =250)

plots = NULL
for (cpg in colnames(df)[-c(dim(df)[2],dim(df)[2]-1)]) {
        plots[[cpg]] = ggplot(df, aes_string("Dx", cpg)) + geom_boxplot(outlier.shape = NA) + geom_jitter(aes(color = df$BH))
        }

grid.arrange(plots[[1]],plots[[2]],plots[[3]],plots[[4]],
             plots[[5]],plots[[6]],plots[[7]],plots[[8]],
             plots[[9]],plots[[10]],plots[[11]],plots[[12]],
             plots[[13]],plots[[14]],plots[[15]],plots[[16]],
            ncol = 4)
dev.off()
```


# Dx Gender and interaction assocaiated DMPs
```{r}
# # Dx Gender interaction
# pData(GRsetPbmcFunFlt)$ibsGenInt <- interaction(pData(GRsetPbmcFunFlt)$Dx,pData(GRsetPbmcFunFlt)$Gender)
# pData(GRsetPbmcFunFlt)$ibsGenInt <- gsub("\\.","_", pData(GRsetPbmcFunFlt)$ibsGenInt)
# dmpsIntPbmc <- DMPsInt(GRsetPbmcFunFlt,factor(pData(GRsetPbmcFunFlt)$ibsGenInt), ann450kSub)

mVals_c <- getM(GRsetColFunFlt)
head(mVals_c[,1:5])
betaVals_c <- getBeta(GRsetColFunFlt)
head(betaVals_c[,1:5])

# factor of interest
Dx <- factor(pData(GRsetColFunFlt)$Dx)
Gender <- factor(pData(GRsetColFunFlt)$Gender)

# variable to correct for- coupld potenitally correct for age, but leave it out

# use the above to create a design matrix
design <- model.matrix(~ 0 + Dx:Gender, data=pData(GRsetColFunFlt))
colnames(design) <- c("HC_F", "IBS_F", "HC_M", "IBS_M")

# fit the linear model 
fit <- lmFit(mVals_c, design)

# create a contrast matrix for specific comparisons
contMatrix <- makeContrasts(IBS_M-HC_M, IBS_F-HC_F,IBS_M-IBS_F, HC_M-HC_F,   levels=design)
contMatrix

# fit the contrasts
fit2 <- contrasts.fit(fit, contMatrix)
fit2 <- eBayes(fit2)

# look at the numbers of DM CpGs at FDR < 0.05
summary(decideTests(fit2))
# 
#       IBS_M - HC_M IBS_F - HC_F IBS_M - IBS_F HC_M - HC_F
# -1            0            0         381         202
# 0        420257       420257        419626      420002
# 1            0            0         250          53

# get the table of results for the first contrast (naive - rTreg)
ann450kSub <- hm450annotations[match(rownames(mVals_c),hm450annotations$Name),
                      c(1:4,12:19,24:ncol(hm450annotations))]
DMPs_IBS_M_HC_M <-  topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)
DMPs_IBS_M_HC_M$meanDiff <- rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "IBS_M"))])-rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "HC_M"))])
DMPs_IBS_M_HC_MSig <- subset(DMPs_IBS_M_HC_M, DMPs_IBS_M_HC_M$adj.P.Val<0.05); dim(DMPs_IBS_M_HC_MSig)

DMPs_IBS_F_HC_F <-  topTable(fit2, num=Inf, coef=2, genelist=ann450kSub)
DMPs_IBS_F_HC_F$meanDiff <- rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "IBS_F"))])-rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "HC_F"))])
DMPs_IBS_F_HC_FSig <- subset(DMPs_IBS_F_HC_F, DMPs_IBS_F_HC_F$adj.P.Val<0.05); dim(DMPs_IBS_F_HC_FSig)

DMPs_IBS_M_IBS_F <-  topTable(fit2, num=Inf, coef=3, genelist=ann450kSub)
DMPs_IBS_M_IBS_F$meanDiff <- rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "IBS_M"))])-rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "IBS_F"))])

DMPs_HC_M_HC_F <- topTable(fit2, num=Inf, coef=4, genelist=ann450kSub)
DMPs_HC_M_HC_F$meanDiff <- rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "HC_M"))])-rowMeans(betaVals_c[,colnames(betaVals_c)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "HC_F"))])

# remove HC Gender probes from IBS
DMPs_IBS_M_IBS_F_sig <- subset(DMPs_IBS_M_IBS_F, DMPs_IBS_M_IBS_F$adj.P.Val<0.05); dim(DMPs_IBS_M_IBS_F_sig)
# [1] 631    33

DMPs_HC_M_HC_F_sig <- subset(DMPs_HC_M_HC_F, DMPs_HC_M_HC_F$adj.P.Val<0.05); dim(DMPs_HC_M_HC_F_sig)
# [1] 255  33

DMPs_HC_M_HC_F_sig <- subset(DMPs_HC_M_HC_F, DMPs_HC_M_HC_F$P.Value<0.05); dim(DMPs_HC_M_HC_F_sig)
# [1] 48568    33


DMPs_IBS_M_IBS_F_sig1 <- DMPs_IBS_M_IBS_F_sig[!row.names(DMPs_IBS_M_IBS_F_sig)%in%row.names(DMPs_HC_M_HC_F_sig),]; dim(DMPs_IBS_M_IBS_F_sig1)

# [1] 224  32
```

```{r}

# ColGenderDMPs <- DMPs(GRsetColFunFlt,factor(pData(GRsetColFunFlt)$Gender), ann450kSub)
# ColGenderDMPs_sig <- subset(ColGenderDMPs, ColGenderDMPs$adj.P.Val<0.05); dim(ColGenderDMPs_sig)

grp <- pData(GRsetColFunFlt)$Gender
M <- getM(GRsetColFunFlt)
B <- getBeta(GRsetColFunFlt)
ColSexDMPs <- dmpFinder(M, pheno=grp, type="categorical", shrinkVar = TRUE)
ColSexDMPs$Bdiff <- (rowMeans(B[row.names(ColSexDMPs),colnames(B)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$Gender == "F"))]))-(rowMeans(B[row.names(ColSexDMPs),colnames(B)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$Gender == "M"))]))
sum(ColSexDMPs$qval < 0.05, na.rm=TRUE)
head(ColSexDMPs)
ann450kSub <- ann450kSub[row.names(ann450kSub)%in%row.names(ColSexDMPs),]
ann450kSub <- ann450kSub[row.names(ColSexDMPs),]
ColSexDMPs <- cbind(ColSexDMPs,ann450kSub)
ColSexDMPsSig <- subset(ColSexDMPs,ColSexDMPs$qval < 0.05 )
```

```{r}
DMPs_IBS_M_IBS_F_sig2 <- DMPs_IBS_M_IBS_F_sig1[!row.names(DMPs_IBS_M_IBS_F_sig1)%in%row.names(ColSexDMPsSig),]; dim(DMPs_IBS_M_IBS_F_sig2)
# [1] 17    33

# DMPs_IBS_M_IBS_F_sig3 <- subset(DMPs_IBS_M_IBS_F_sig2, DMPs_IBS_M_IBS_F_sig2$adj.P.Val<0.01);dim(DMPs_IBS_M_IBS_F_sig3)
# # [1] 4072   32

```

```{r}

# write.table(ColDxDMPs, file="/temp1/res3_functional_norm/methBioconductorWorkflow/colon/sex/ColDxDMPs.csv", sep=",", row.names=FALSE)

# write.table(ColDxDMPs_sig_p001, file="/temp1/res3_functional_norm/methBioconductorWorkflow/colon/sex/ColDxDMPs_sig_p001.csv", sep=",", row.names=FALSE)
```

```{r}
#################### save the results

write.table(ColSexDMPsSig, file="/temp1/res3_functional_norm/methBioconductorWorkflow/colon/sex/ColGenderDMPs.csv", sep=",", row.names=FALSE)

save(ColSexDMPs, file = "C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/sex/ColSexDMPs.rda")

save(fit2, file = "C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/IBS_sex_interaction/fitted_gender_dx_int_Col.rda")

write.table(DMPs_IBS_F_HC_FSig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/IBS_sex_interaction/Col_DMPs_IBS_F_HC_FSig.csv", sep=",", row.names=FALSE)

write.table(DMPs_IBS_M_HC_MSig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/IBS_sex_interaction/Col_DMPs_IBS_M_HC_MSig.csv", sep=",", row.names=FALSE)

write.table(DMPs_IBS_M_IBS_F_sig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/IBS_sex_interaction/DMPs_IBS_M_IBS_F_sig.csv", sep=",", row.names=FALSE)

 write.table(DMPs_HC_M_HC_F_sig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/IBS_sex_interaction/DMPs_HC_M_HC_F_sig.csv", sep=",", row.names=FALSE)

write.table(DMPs_IBS_M_IBS_F_sig2, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/colon/IBS_sex_interaction/DMPs_IBS_M_IBS_F_sig2.csv", sep=",", row.names=FALSE)

# write.table(DMPs_IBS_M_IBS_F_sig3, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/Col/DMPs_IBS_M_IBS_F_sig3.csv", sep=",", row.names=FALSE)

# save(dmpsIntCol , file = "temp/functional_norm/methBioconductorWorkflow/Col/Col_dmpsIntCol.rda")
```







# Dx Gender and interaction assocaiated DMPs

```{r}
# Dx Gender interaction
pData(GRsetColFunFlt)$ibsGenInt <- interaction(pData(GRsetColFunFlt)$Dx,pData(GRsetColFunFlt)$Gender)
pData(GRsetColFunFlt)$ibsGenInt <- gsub("\\.","_", pData(GRsetColFunFlt)$ibsGenInt)
dmpsIntCol <- DMPsInt(GRsetColFunFlt,factor(pData(GRsetColFunFlt)$ibsGenInt), ann450kSub)

mVals_p <- getM(GRsetColFunFlt)
head(mVals_p[,1:5])
betaVals_p <- getBeta(GRsetColFunFlt)
head(betaVals_p[,1:5])

# factor of interest
Dx <- factor(pData(GRsetColFunFlt)$Dx)
Gender <- factor(pData(GRsetColFunFlt)$Gender)

# variable to correct for- coupld potenitally correct for age, but leave it out

# use the above to create a design matrix
design <- model.matrix(~ 0 + Dx:Gender, data=pData(GRsetColFunFlt))
colnames(design) <- c("HC_F", "IBS_F", "HC_M", "IBS_M")

# fit the linear model 
fit <- lmFit(mVals_p, design)

# create a contrast matrix for specific comparisons
contMatrix <- makeContrasts(IBS_M-HC_M, IBS_F-HC_F,IBS_M-IBS_F, HC_M-HC_F,   levels=design)
contMatrix

# fit the contrasts
fit2 <- contrasts.fit(fit, contMatrix)
fit2 <- eBayes(fit2)

# look at the numbers of DM CpGs at FDR < 0.05
summary(decideTests(fit2))
# 
#   IBS_M - HC_M IBS_F - HC_F IBS_M - IBS_F HC_M - HC_F
# -1            0            0           381         202
# 0        420257       420257        419626      420002
# 1             0            0           250          53

# get the table of results for the first contrast (naive - rTreg)
ann450kSub <- hm450annotations[match(rownames(mVals_p),hm450annotations$Name),
                      c(1:4,12:19,24:ncol(hm450annotations))]
DMPs_IBS_M_HC_M <-  topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)
DMPs_IBS_M_HC_M$meanDiff <- rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "IBS_M"))])-rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "HC_M"))])

DMPs_IBS_F_HC_F <-  topTable(fit2, num=Inf, coef=2, genelist=ann450kSub)
DMPs_IBS_F_HC_F$meanDiff <- rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "IBS_F"))])-rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "HC_F"))])

DMPs_IBS_M_IBS_F <-  topTable(fit2, num=Inf, coef=3, genelist=ann450kSub)
DMPs_IBS_M_IBS_F$meanDiff <- rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "IBS_M"))])-rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "IBS_F"))])

DMPs_HC_M_HC_F <- topTable(fit2, num=Inf, coef=4, genelist=ann450kSub)
DMPs_HC_M_HC_F$meanDiff <- rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "HC_M"))])-rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetColFunFlt), pData(GRsetColFunFlt)$ibsGenInt ==  "HC_F"))])

# remove HC Gender probes from IBS
DMPs_IBS_M_IBS_F_sig <- subset(DMPs_IBS_M_IBS_F, DMPs_IBS_M_IBS_F$adj.P.Val<0.05); dim(DMPs_IBS_M_IBS_F_sig)
# [1] 631    33

DMPs_HC_M_HC_F_sig <- subset(DMPs_HC_M_HC_F, DMPs_HC_M_HC_F$adj.P.Val<0.05); dim(DMPs_HC_M_HC_F_sig)
# [1] 255  33

DMPs_HC_M_HC_F_sig <- subset(DMPs_HC_M_HC_F, DMPs_HC_M_HC_F$P.Value<0.05); dim(DMPs_HC_M_HC_F_sig)
# [1] 48568    33


DMPs_IBS_M_IBS_F_sig1 <- DMPs_IBS_M_IBS_F_sig[!row.names(DMPs_IBS_M_IBS_F_sig)%in%row.names(DMPs_HC_M_HC_F_sig),]; dim(DMPs_IBS_M_IBS_F_sig1)

# [1] 224  33
```

```{r}

ColGenderDMPs <- DMPs(GRsetColFunFlt,factor(pData(GRsetColFunFlt)$Gender), ann450kSub)
ColGenderDMPs_sig <- subset(ColGenderDMPs, ColGenderDMPs$adj.P.Val<0.05); dim(ColGenderDMPs_sig)
# [1] 2278   33
```

```{r}
DMPs_IBS_M_IBS_F_sig2 <- DMPs_IBS_M_IBS_F_sig1[!row.names(DMPs_IBS_M_IBS_F_sig1)%in%row.names(ColGenderDMPs_sig),]; dim(DMPs_IBS_M_IBS_F_sig2)
# [1] 28    33

DMPs_IBS_M_IBS_F_sig3 <- subset(DMPs_IBS_M_IBS_F_sig2, DMPs_IBS_M_IBS_F_sig2$adj.P.Val<0.01);dim(DMPs_IBS_M_IBS_F_sig3)
# [1] 0   33

# DMPs_IBS_M_IBS_F_sig3 <- subset(DMPs_IBS_M_IBS_F_sig2, DMPs_IBS_M_IBS_F_sig2$meanDiff>=0.02|DMPs_IBS_M_IBS_F_sig2$meanDiff< -0.02); dim(DMPs_IBS_M_IBS_F_sig3)
# 
# save(DMPs_IBS_M_IBS_F_sig2, file = "C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/colon/DMPs_IBS_M_IBS_F_sig2.rda")
# 
# write.csv(DMPs_IBS_M_IBS_F_sig3, file = "C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/colon/DMPs_IBS_M_IBS_F_sig2_28.csv")

```

```{r}
## same results from making an interaction variable.

dmpsIntCol <- read.csv("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/Col/dmpsIntCol.csv", row.names = 4)

ibs_m_f_sig <- subset(dmpsIntCol[[3]], dmpsIntCol[[3]]$adj.P.Val<0.05); dim(ibs_m_f_sig)
# [1] 40280    33
hc_m_f_sig <- subset(dmpsIntCol[[4]], dmpsIntCol[[4]]$adj.P.Val<0.05); dim(hc_m_f_sig)
# # [1] 453  33 
ibs_m_f_sig1 <- ibs_m_f_sig[!row.names(ibs_m_f_sig)%in%row.names(hc_m_f_sig),]; dim(ibs_m_f_sig1)
# # [1] 39990    33

## very few gender differences in HC compared to IBS.use p value instead to remove as many gender assocaited probes as possible.
hc_m_f_sig <- subset(dmpsIntCol[[4]], dmpsIntCol[[4]]$P.Value<0.05); dim(hc_m_f_sig)
# [1] 30448    33
ibs_m_f_sig1 <- ibs_m_f_sig[!row.names(ibs_m_f_sig)%in%row.names(hc_m_f_sig),]; dim(ibs_m_f_sig1)
# [1] 37363    33

# IBS HC differences within 1. Women 2. men
f_ibs_hc_sig <- subset(dmpsIntCol[[1]], dmpsIntCol[[1]]$adj.P.Val<0.05); dim(f_ibs_hc_sig)
# [1]  2 33
m_ibs_hc_sig<- subset(dmpsIntCol[[2]], dmpsIntCol[[2]]$adj.P.Val<0.05); dim(m_ibs_hc_sig)
# [1] 15 33

## try to remove general gender differences
ibs_m_f_sig2 <- ibs_m_f_sig[!row.names(ibs_m_f_sig)%in%row.names(ColGenderDMPs_sig),]; dim(ibs_m_f_sig2)
# [1] 31759    33 
ibs_m_f_sig3 <- ibs_m_f_sig2[!row.names(ibs_m_f_sig2)%in%row.names(hc_m_f_sig),]; dim(ibs_m_f_sig3)
# [1] 30821    33 ### final

ibs_m_f_sig3fdr01 <- subset(ibs_m_f_sig3, ibs_m_f_sig3$adj.P.Val<0.01); dim(ibs_m_f_sig3fdr01)
# [1] 5788   33

# ibs_m_f_sig3fdr01md <- subset(ibs_m_f_sig3fdr01, ibs_m_f_sig3fdr01$meanDiff >0.3 | ibs_m_f_sig3fdr01$meanDiff < -0.3); dim(ibs_m_f_sig3fdr01md )

```

```{r setup}
# Set working directory outside the chunks
knitr::opts_knit$set(root.dir = normalizePath("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/"))
```

```{r}
#################### save the results

write.table(ColDxDMPs, file="temp/functional_norm/methBioconductorWorkflow/colon/ColDxDMPs.csv", sep=",", row.names=FALSE)

write.table(ColDxDMPs_sig_p001, file="temp/functional_norm/methBioconductorWorkflow/colon/ColDxDMPs_sig_p001.csv", sep=",", row.names=FALSE)

write.table(ColGenderDMPs, file="temp/functional_norm/methBioconductorWorkflow/colon/ColGenderDMPs.csv", sep=",", row.names=FALSE)

write.table(dmpsIntCol, file = "temp/functional_norm/methBioconductorWorkflow/colon/dmpsIntCol.csv", sep=",", row.names=FALSE)

write.table(f_ibs_hc_sig, file="temp/functional_norm/methBioconductorWorkflow/colon/Col_f_ibs_hc_sig.csv", sep=",", row.names=FALSE)

write.table(m_ibs_hc_sig, file="temp/functional_norm/methBioconductorWorkflow/colon/Col_m_ibs_hc_sig.csv", sep=",", row.names=FALSE)

write.table(ibs_m_f_sig2, file="temp/functional_norm/methBioconductorWorkflow/colon/Col_ibs_m_f_sig2.csv", sep=",", row.names=FALSE)

# write.table(ibs_m_f_sig3, file="temp/functional_norm/methBioconductorWorkflow/colon/Col_ibs_m_f_sig_final.csv", sep=",", row.names=FALSE)

write.table(ColGenderDMPs_sig, file="temp/functional_norm/methBioconductorWorkflow/colon/ColGenderDMPs_sig.csv", sep=",", row.names=FALSE)

write.table(ColDxDMPs_sig_p05, file="temp/functional_norm/methBioconductorWorkflow/colon/ColDxDMPs_sig_p05.csv", sep=",", row.names=FALSE)


save(dmpsIntCol , file = "temp/functional_norm/methBioconductorWorkflow/colon/Col_dmpsIntCol.rda")
```


