---
title: "pbmc_part_Methyl_paper_1_2_2018"
author: "Swapna Mahurkar-Joshi"
date: "January 26, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Genome-wide DNA Methylation and Gene Expression of Colonic Mucosa and Peripheral Blood Mononuclear Cells Reveal Stress Associated and Neuronal Genes to be Epigenetically Deregulated in IBS

This is the script for the data and the figures in the paper.

# Libraries

```{r echo=FALSE, message=FALSE, warning=FALSE}
# source("https://bioconductor.org/biocLite.R")
# biocLite("MASS")
library(ggplot2)
library(scater)
library(limma)
library(minfi)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(IlluminaHumanMethylation450kmanifest)
library(RColorBrewer)
library(missMethyl)
library(matrixStats)
library(Gviz)
library(DMRcate)
library(stringr)
library(methylationArrayAnalysis)
library(FDb.InfiniumMethylation.hg19)
library(RColorBrewer)
library(Gviz)
library(GenomicRanges)
library(DMRcate)
library(knitr)
```
# Set the working directory

```{r setup}
# Set working directory outside the chunks
knitr::opts_knit$set(root.dir = normalizePath("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/"))
```

```{r}
load("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/data1/consistant_data/pbmc_meth_level4.rda")
ann450kSub <- pbmc_meth_level4[["ann450kSub"]] 
metDat_p   <- pbmc_meth_level4[["metDat_p"]]; dim(metDat_p)
pbmc_betas <- pbmc_meth_level4[["pbmc_betas"]]
rm(pbmc_meth_level4)

metDat_p$BH <-  gsub("U", "M", metDat_p$BH)
ibs    <- subset(metDat_p, metDat_p$Group==2)
hc     <- subset(metDat_p, metDat_p$Group==1)
```

# PBMCs
##Table 1: Demographics for PBMC data

```{r}
table1 <- matrix(NA, nrow = 16, ncol = 3)

table1[c(1,2,5:16), 1] <- round(apply(ibs[ ,c(4,5,6,7,8,9,14,15,16, 17,18,21,25,27)], 2, mean, na.rm = TRUE), 2)
table1[c(1,2,5:16), 2] <- round(apply(hc[ ,c(4,5,6,7,8,9,14,15,16,17,18,21,25,27)], 2, mean, na.rm = TRUE), 2)

table2 <- matrix(NA, nrow = 16, ncol = 3)
table2[c(1,2,5:16), 1] <- round(apply(ibs[ ,c(4,5,6,7,8,9,14,15,16, 17,18,21,25,27)], 2, sd, na.rm = TRUE), 2)
table2[c(1,2,5:16), 2] <- round(apply(hc[ ,c(4,5,6,7,8,9,14,15,16,17,18,21,25,27)], 2, sd, na.rm = TRUE), 2)
row.names(table2)[c(1,2,5:16)] <- colnames(hc[, c(4,5,6,7,8,9,14,15,16,17,18,21,25,27)])

table1[3,1] <- round(table(ibs$Sex)[2]/(table(ibs$Sex)[2]+table(ibs$Sex)[1])*100,2)
table1[3,2] <-round(table(hc$Sex)[2]/(table(hc$Sex)[2]+table(hc$Sex)[1])*100,2)

row.names(table1)[c(1,2,5:16)] <- colnames(hc[, c(4,5,6,7,8,9,14,15,16,17,18,21,25,27)])
row.names(table1)[3]<-"Female %"
row.names(table1)[4]<-"Bowel Habit %"

sd(ibs$Age, na.rm = TRUE)
sd(hc$Age, na.rm = TRUE)
```

```{r}
metDat_p$Group <- as.factor(metDat_p$Group)
comb1          <- expand.grid(names(metDat_p[,c(4,5,6,7,8,9,14,15,16,17,18,25,27)]),"Group")

models     <- lapply(paste(comb1[,1],comb1[,2], sep = "~"), formula)
res.models <- lapply(models, FUN = function(x) {t.test(formula = x, data = metDat_p)})
# names(res.models) <- paste(comb1[,1],comb1[,2], sep = "~")
table1[c(1:2,5:13,15:16),3] <- round(t(data.frame(lapply(res.models, FUN = t_p))),5)
colnames(table1) <- c("IBS (N=109)","Healthy Controls (N=36)", "t.test p-value")
table2 <- matrix(as.numeric(as.character(table1)), nrow = 16, ncol = 3)
row.names(table2) <- row.names(table1)
colnames(table2) <- colnames(table1)
table2 <-  as.data.frame(table2)
table2[c(5,6,14),c(2,3)] <- "-"
# Fisher test for sex and bowel habits
fisher.test(metDat_p$Group, metDat_p$Sex)
table2[3,3] <- 0.168
ibsBHper    <- c(round(table(ibs$BH)[1]/dim(ibs)[1]*100,2),round(table(ibs$BH)[2]/dim(ibs)[1]*100,2),round(table(ibs$BH)[3]/dim(ibs)[1]*100,2),round(table(ibs$BH)[4]/dim(ibs)[1]*100,2))
table2[3,1] <- paste(paste("IBS-C",ibsBHper[1],sep = "="),paste("IBS-D",ibsBHper[2],sep = "="),paste("IBS-M",ibsBHper[3],sep = "="), sep=",")

kable(table2,align=c(rep('c', 16)), caption = "Table1: Clinical Characteristics of IBS Patients and HCs")
```

## PBMCs:DMPs between IBS and HCs

```{r}
pbmcDxDMPs <- dmpFinder(getM(GRsetPbmcFunFlt),pData(GRsetPbmcFunFlt)$Dx, type = "categorical", qCutoff = 1, shrinkVar = TRUE)



```

```{r}
pbmcDxDMPs_sig <- subset(pbmcDxDMPs, pbmcDxDMPs$qval<0.05); dim(pbmcDxDMPs_sig)
# [1]  0 33
pbmcDxDMPs_sig_p05 <- subset(pbmcDxDMPs, pbmcDxDMPs$pval<0.05); dim(pbmcDxDMPs_sig_p05)
# [1] 13617    33
pbmcDxDMPs_sig_p001 <- subset(pbmcDxDMPs, pbmcDxDMPs$pval<0.001); dim(pbmcDxDMPs_sig_p001)
# [1] 179  33
cpgs <- rownames(pbmcDxDMPs_sig_p001)[1:4]
par(mfrow=c(2,2))
plotCpg(getBeta(GRsetPbmcFunFlt), cpg=cpgs, pheno=pData(GRsetPbmcFunFlt)$Dx)
ggplot(aes())
```

```{r setup}
# Set working directory outside the chunks
knitr::opts_knit$set(root.dir = normalizePath("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/Methyl_paper_1_2_2018"))
```

#  PBMCs:DMRs between IBS and HCs

```{r}
targets <- pData(GRsetPbmcFunFlt)
dx <- factor(targets$Dx)
design <- model.matrix(~0+dx, data=targets)
colnames(design) <- levels(dx)
contMatrix <- makeContrasts(IBS-HC,   levels=design)
contMatrix
```

```{r}
myAnnotation <- cpg.annotate(object = getM(GRsetPbmcFunFlt), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "IBS - HC", arraytype = "450K")
str(myAnnotation)
DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]
```

```{r}
results.ranges <- extractRanges(DMRs, genome = "hg19")
write.csv(results.ranges, file = "PbmcDMRs_p001.csv")
save(results.ranges, file = "PbmcDMRs_p001.rda")
```

```{r}
# draw the plot for the top DMR
# set up the grouping variables and colors
pal <- brewer.pal(3,"Dark2")
groups <- pal[1:length(unique(targets$Dx))]
names(groups) <- levels(factor(targets$Dx))
cols <- groups[as.character(factor(targets$Dx))]
samps <- 1:nrow(targets)
png("temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDxDMR1.png", height = 2000, width = 2000, res = 250)
DMR.plot(ranges=results.ranges, dmr=1, CpGs=getBeta(GRsetPbmcFunFlt), phen.col=cols, what = "Beta",      arraytype = "450K", pch=16, toscale=TRUE, plotmedians=TRUE,   genome="hg19", samps=samps)
dev.off()
```
## PBMCs:DMPs and DMRs between IBS-C and HCs; IBS-D and HCs; IBS-M and HCs:
## PBMCs:IBS-Sex interaction: 
## PBMCs:IBS-Sex interaction ANOVA:
## PBMCs:Men-Women differences within IBS
## PBMCs:Men-Women differences within HCs
## PBMCs:Stress related genes: 
## PBMCs:HPA axis: 
## PBMCs:GR inducible and GR repressed: 
## PBMCs:Gene expression changes associated with genes associated with IBS, BH subtypes or IBS-Sex interactions:
## PBMCs:Functional annotation of the genes:
######################################################################################
# Dx associated DMPs in PBMCs
######################################################################################

```{r}
load("temp/functional_norm/methBioconductorWorkflow/GRsetPbmcColFunFlt.Rda")
load("temp/hm450annotations.rda")
ann450kSub <- hm450annotations[match(featureNames(GRsetColFunFlt),hm450annotations$Name),c(1:4,12:19,24:ncol(hm450annotations))] 
# PbmcDxDMPs <- DMPs(GRsetPbmcFunFlt,factor(pData(GRsetPbmcFunFlt)$Dx), ann450kSub)
# pbmcDxDMP <- dmpFinder(getM(GRsetPbmcFunFlt),pData(GRsetPbmcFunFlt)$Dx, type = "categorical", qCutoff = 1, shrinkVar = TRUE)
```

```{r}
library(minfiData)
if (require(minfiData)) {

grp <- pData(GRsetPbmcFunFlt)$Dx
M <- getM(GRsetPbmcFunFlt)
B <- getBeta(GRsetPbmcFunFlt)
PbmcDxDMPs <- dmpFinder(M, pheno=grp, type="categorical", shrinkVar = TRUE)
PbmcDxDMPs$Bdiff <- rowMeans(B[row.names(PbmcDxDMPs),colnames(B)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$Dx == "IBS"))])-rowMeans(B[row.names(PbmcDxDMPs),colnames(B)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$Dx == "HC"))])
sum(PbmcDxDMPs$qval < 0.05, na.rm=TRUE)
head(PbmcDxDMPs)
ann450kSub <- ann450kSub[row.names(ann450kSub)%in%row.names(PbmcDxDMPs),]
ann450kSub <- ann450kSub[row.names(PbmcDxDMPs),]
PbmcDxDMPs <- cbind(PbmcDxDMPs,ann450kSub)
}


save(PbmcDxDMPs, file= "C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDxDMPs.rda")
write.csv(PbmcDxDMPs, file= "C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDxDMPs.csv")
# PbmcDxDMPs <- read.csv("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDxDMPs.csv", row.names = 1)
PbmcDxDMPs_sig <- subset(PbmcDxDMPs, PbmcDxDMPs$qval<0.05); dim(PbmcDxDMPs_sig)
# [1]  0 31
PbmcDxDMPs_sig_p05 <- subset(PbmcDxDMPs, PbmcDxDMPs$pval<0.05); dim(PbmcDxDMPs_sig_p05)
# [1] 13617    33
PbmcDxDMPs_sig_p001 <- subset(PbmcDxDMPs, PbmcDxDMPs$pval<0.001); dim(PbmcDxDMPs_sig_p001)
# [1] 179  33
write.csv(PbmcDxDMPs_sig_p001, file= "C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDxDMPs_sig_p001.csv")

betas_Pbmc <- getBeta(GRsetPbmcFunFlt)
design <- model.matrix(~0+pData(GRsetPbmcFunFlt)$Dx, data=pData(GRsetPbmcFunFlt))
  colnames(design) <- levels(factor(pData(GRsetPbmcFunFlt)$Dx))
  fit <- lmFit(betas_Pbmc, design)
  contMatrix <- makeContrasts(IBS-HC ,levels=design)
  fit2 <- contrasts.fit(fit, contMatrix)
  fit2 <- eBayes(fit2,robust=T, trend=T)
  tt <- topTable(fit2, coef=1,number=Inf, genelist=ann450kSub)
  
  md1 <- rowMeans(betas_Pbmc[,colnames(betas_Pbmc)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$Dx ==  levels(as.factor(pData(GRsetPbmcFunFlt)$Dx))[2]))]) - rowMeans(betas_Pbmc[,colnames
(betas_Pbmc)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$Dx ==  levels(as.factor(pData(GRsetPbmcFunFlt)$Dx))[1]))])
  tt$meanDiff <- md1
library(limma)
library(missMethyl)
gst.KEGG <- gometh(sig.cpg=rownames(PbmcDxDMPs), all.cpg=rownames(betas_Pbmc),  prior.prob = T)
gst.KEGG <- gst.KEGG[order(gst.KEGG$P.DE),]
gst.KEGG  <- gst.KEGG[gst.KEGG$FDR<0.05,]

# ann <- missMethyl:::.flattenAnn(fit2)
# m <- match(rownames(fit2), ann$cpg)
# EntrezID <- ann$entrezid[m]
# n <- table(EntrezID)
# k <- kegga(fit2, geneid=EntrezID, covariate=n[EntrezID])
# topKEGG(k)



```

```{r setup}
# Set working directory outside the chunks
knitr::opts_knit$set(root.dir = normalizePath("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/"))
```

# DMRs
```{r}

library(Gviz)
library(GenomicRanges)
library(DMRcate)

targets <- pData(GRsetPbmcFunFlt)
dx <- factor(targets$Dx)
design <- model.matrix(~0+dx, data=targets)
colnames(design) <- levels(dx)
contMatrix <- makeContrasts(IBS-HC,   levels=design)
contMatrix

myAnnotation <- cpg.annotate(object = getM(GRsetPbmcFunFlt), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "IBS - HC", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : Factor w/ 420257 levels "cg00000029","cg00000108",..: 222425 374552 248925 18571 277135 194037 362374 266425 100898 55258 ...
#  $ stat  : num [1:420257] 1.095 -2.231 -0.322 0.212 -0.347 ...
#  $ CHR   : Factor w/ 22 levels "chr1","chr10",..: 1 1 1 1 1 1 1 1 1 1 ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 9.02e-03 -1.32e-02 -1.72e-03 4.78e-05 -8.13e-04 ...
#  $ indfdr: num [1:420257] 1 1 1 1 1 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"


DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]
# 
#                       coord no.cpgs       minfdr Stouffer  maxbetafc meanbetafc
# 12  chr20:57427274-57427942      18 8.791846e-05        1 0.04018264 0.01837412
# 11  chr20:43935222-43935551       9 2.333466e-05        1 0.04273920 0.02485073
# 10  chr19:57630446-57630691       8 8.025709e-04        1 0.03009786 0.01785044
# 1    chr1:24229232-24229682       5 8.548694e-06        1 0.03850575 0.03519281
# 5       chr16:450562-451040       5 8.791846e-05        1 0.08800632 0.05547127
# 8   chr17:80393124-80393666       5 1.889939e-06        1 0.02761253 0.01341668
# 6     chr16:3079708-3079953       4 8.025709e-04        1 0.03098642 0.02761540
# 2   chr13:58205001-58205111       3 1.404496e-05        1 0.02586554 0.01196997
# 7   chr17:72270155-72270444       3 2.248281e-04        1 0.04970472 0.02936569
# 13 chr6:170337747-170337897       3 3.914138e-04        1 0.01947245 0.01364406
# 4   chr15:75019302-75019376       2 8.025709e-04        1 0.03146319 0.01589445
# 9   chr18:43418829-43419102       2 6.145642e-04        1 0.04767722 0.04049430

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
save(results.ranges, file = "temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDMRs_p001.rda")
write.csv(results.ranges, file = "temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDMRs_p001.csv")
# set up the grouping variables and colors
pal <- brewer.pal(3,"Dark2")
groups <- pal[1:length(unique(targets$Dx))]
names(groups) <- levels(factor(targets$Dx))
cols <- groups[as.character(factor(targets$Dx))]
samps <- 1:nrow(targets)

# draw the plot for the top DMR
png("temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDxDMR1.png", height = 2000, width = 2000, res = 250)
DMR.plot(ranges=results.ranges, dmr=1, CpGs=getBeta(GRsetPbmcFunFlt), phen.col=cols, what = "Beta",      arraytype = "450K", pch=16, toscale=TRUE, plotmedians=TRUE,   genome="hg19", samps=samps)
dev.off()

```

```{r}
## PCDH17 DMR
pcdh17SigCpgs <- PbmcDxDMPs_sig_p05[PbmcDxDMPs_sig_p05$nearestGeneSymbol=="PCDH17",] # 3rd in in opp direction
pcdh17SigCpgsPlusStr <- subset(pcdh17SigCpgs, pcdh17SigCpgs$strand == "+")
pcdh17SigCpgsNegStr <- subset(pcdh17SigCpgs, pcdh17SigCpgs$strand == "-")

pcdh17SigCpgs <- pcdh17SigCpgsPlusStr
pcdh17SigCpgs <- pcdh17SigCpgs[order(pcdh17SigCpgs$pos),]
betas_pcdh17dmr <- betas_Pbmc[row.names(betas_Pbmc)%in%row.names(pcdh17SigCpgs),]
betas_pcdh17dmr <- betas_Pbmc[row.names(pcdh17SigCpgs),]
dx <- as.factor(ifelse(as.factor(pData(GRsetPbmcFunFlt)$Dx)=="HC","#0000ff","#ff0000"))
grObj <- makeGRangesFromDataFrame(data.frame(chr = pcdh17SigCpgs$chr, start = pcdh17SigCpgs$pos, end = pcdh17SigCpgs$pos+1, strand = pcdh17SigCpgs$strand, score = pcdh17SigCpgs$meanDiff))


atrack <- AnnotationTrack(grObj, name = "CpG")
plotTracks(atrack)
chromosome(atrack) <- 'chr13'

gtrack <- GenomeAxisTrack()
plotTracks(list(gtrack, atrack))

# genome : "hg19" 
# gen<-genome(grObj)

gen = "hg19" 
chr <- as.character(unique(seqnames(grObj)))

# Ideogram track
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
plotTracks(list(itrack, gtrack, atrack))
# geneModels <- data.frame(c(rep("chr13",2)), c(58206655,58205944),c(58303445,58299695),c(96790,93751), c("+","+"),c("protein coding","Nonsense mediated decay"),c("ENSG00000118946" , "ENSG00000118946"),c("ENSG00000118946" , "ENSG00000118946") ,c("ENST00000377918","ENST00000484979"),c("PCDH17","PCDH17"))
# colnames(geneModels) <- c( "chromosome",    "start",      "end", "width", "strand", "feature","gene",  "exon",      "transcript",     "symbol")
# 
# grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr, name = "Gene Model")
# chromosome(grtrack) <- 'chr13'
# plotTracks(list(itrack, gtrack, atrack, grtrack))

##data track
 # coords <- c(rep(58205001,145),rep(58205045,145), rep(58209133,145))
# dat <- c(t(betas_pcdh17dmr)[,1], t(betas_pcdh17dmr)[,2],t(betas_pcdh17dmr)[,3] ) 
mcols(grObj) <- betas_pcdh17dmr
 # Dx <- ifelse(dx=="IBS","red","blue")
dx <- factor(targets$Dx)

dtrack <- DataTrack(grObj,groups = as.vector(dx),type= c("b"),aggregateGroups = TRUE,
           aggregation = "mean")
plotTracks(dtrack, groups = as.vector(dx), type= "b", aggregateGroups = TRUE,
           aggregation = "max")
plotTracks(list(itrack, gtrack, atrack,dtrack))


```

```{r}
## GNAS DMR
GNASSigCpgs <- PbmcDxDMPs_sig_p05[PbmcDxDMPs_sig_p05$nearestGeneSymbol=="GNAS",]
GNASSigCpgsPlusStr <- subset(GNASSigCpgs, GNASSigCpgs$strand == "+")
GNASSigCpgsNegStr <- subset(GNASSigCpgs, GNASSigCpgs$strand == "-")

GNASSigCpgs <- GNASSigCpgsNegStr
GNASSigCpgs <- GNASSigCpgs[order(GNASSigCpgs$pos),]
betas_NR4A2dmr <- getBeta(GRsetPbmcFunFltCD)[row.names(getBeta(GRsetPbmcFunFltCD))%in%row.names(GNASSigCpgs),]
betas_GNASdmr <- betas_Pbmc[row.names(betas_Pbmc)%in%row.names(GNASSigCpgs),]
dx <- as.factor(pData(GRsetPbmcFunFlt)$Dx)
grObj <- makeGRangesFromDataFrame(data.frame(chr = GNASSigCpgs$chr, start = GNASSigCpgs$pos, end = GNASSigCpgs$pos+1, strand = GNASSigCpgs$strand, score = GNASSigCpgs$meanDiff))


atrack <- AnnotationTrack(grObj, name = "CpG")
plotTracks(atrack)
chromosome(atrack) <- 'chr20'

gtrack <- GenomeAxisTrack()
plotTracks(list(gtrack, atrack))

# genome : "hg19" 
# gen<-genome(grObj)

gen = "hg19" 
chr <- as.character(unique(seqnames(grObj)))

# Ideogram track
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
plotTracks(list(itrack, gtrack, atrack))
# geneModels <- data.frame(c(rep("chr20",2)), c(58206655,58205944),c(58303445,58299695),c(96790,93751), c("+","+"),c("protein coding","Nonsense mediated decay"),c("ENSG00000118946" , "ENSG00000118946"),c("ENSG00000118946" , "ENSG00000118946") ,c("ENST00000377918","ENST00000484979"),c("PCDH17","PCDH17"))
# colnames(geneModels) <- c( "chromosome",    "start",      "end", "width", "strand", "feature","gene",  "exon",      "transcript",     "symbol")

# grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr, name = "Gene Model")
# chromosome(grtrack) <- 'chr13'
# plotTracks(list(itrack, gtrack, atrack, grtrack))

##data track
 # coords <- c(rep(58205001,145),rep(58205045,145), rep(58209133,145))
mcols(grObj) <- betas_GNASdmr
 # Dx <- ifelse(dx=="IBS","red","blue")

dtrack <- DataTrack(grObj,groups = as.vector(dx),type= c("b"),aggregateGroups = TRUE,
           aggregation = "mean")
# 
# plotTracks(dtrack, groups = as.vector(bh), type= "b", aggregateGroups = TRUE,
#            aggregation = "max")
plotTracks(list(itrack, gtrack, atrack, dtrack))
```

######################################################################################
# BH associated DMPs PBMC
########################################################################################

```{r}
GRsetPbmcFunFltCD <- GRsetPbmcFunFlt[,colnames(GRsetPbmcFunFlt)%in%c(row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH == "C"),]), row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH=="D"),]))]

# pbmcBhCDDMP <- dmpFinder(getM(GRsetPbmcFunFltCD),pData(GRsetPbmcFunFltCD)$BH, type = "categorical", qCutoff = 1, shrinkVar = TRUE)

library(minfiData)
if (require(minfiData)) {

grp <- pData(GRsetPbmcFunFltCD)$BH
M <- getM(GRsetPbmcFunFltCD)
B <- getBeta(GRsetPbmcFunFltCD)
PbmcBhDMPs <- dmpFinder(M, pheno=grp, type="categorical", shrinkVar = TRUE)
PbmcBhDMPs$Bdiff <- rowMeans(B[row.names(PbmcBhDMPs),colnames(B)%in%row.names(subset(pData(GRsetPbmcFunFltCD), pData(GRsetPbmcFunFltCD)$Dx == "C"))])-rowMeans(B[row.names(PbmcBhDMPs),colnames(B)%in%row.names(subset(pData(GRsetPbmcFunFltCD), pData(GRsetPbmcFunFltCD)$BH == "D"))])
sum(PbmcBhDMPs$qval < 0.05, na.rm=TRUE)
head(PbmcBhDMPs)
ann450kSub <- ann450kSub[row.names(ann450kSub)%in%row.names(PbmcBhDMPs),]
ann450kSub <- ann450kSub[row.names(PbmcBhDMPs),]
PbmcBhDMPs <- cbind(PbmcBhDMPs,ann450kSub)
}

```



```{r}
mVals_p <- getM(GRsetPbmcFunFlt)
head(mVals_p[,1:5])
betaVals_p <- getBeta(GRsetPbmcFunFlt)
head(betaVals_p[,1:5])

# factor of interest
bh <- factor(pData(GRsetPbmcFunFlt)$BH)

# variable to correct for- coupld potenitally correct for age, but leave it out

# use the above to create a design matrix
design <- model.matrix(~ 0 + bh, data=pData(GRsetPbmcFunFlt))
colnames(design) <- c("bhC", "bhD", "bhM", "bhN")

# fit the linear model 
fit <- lmFit(mVals_p, design)

# create a contrast matrix for specific comparisons
contMatrix <- makeContrasts(bhC-bhN, bhD-bhN,bhM-bhN, bhC-bhD,bhM-bhD, bhM-bhC,levels=design)
contMatrix

# fit the contrasts
fit2 <- contrasts.fit(fit, contMatrix)
fit2 <- eBayes(fit2,trend=TRUE)

# look at the numbers of DM CpGs at FDR < 0.05
summary(decideTests(fit2))
#  bhC - bhN bhD - bhN bhM - bhN bhC - bhD bhM - bhD bhM - bhC
# -1         0         0         0         0         0         0
# 0     420257    420257    420257    420257    420257    420257
# 1          0         0         0         0         0         0

# get the table of results for the first contrast 
DMPs_CN <-  topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)
DMPs_CN$meanDiff <- (rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$BH ==  "C"))]))-(rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$BH ==  "N"))]))
DMPs_CNp001 <- subset(DMPs_CN, DMPs_CN$P.Value<0.001); dim(DMPs_CNp001)
# [1] 150  33

DMPs_DN <-  topTable(fit2, num=Inf, coef=2, genelist=ann450kSub)
DMPs_DN$meanDiff <- (rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$BH ==  "D"))]))-(rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$BH ==  "N"))]))
DMPs_DNp001 <- subset(DMPs_DN, DMPs_DN$P.Value<0.001); dim(DMPs_DNp001)
# [1] 229  33

DMPs_MN <-  topTable(fit2, num=Inf, coef=3, genelist=ann450kSub)
DMPs_MN$meanDiff <- (rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$BH ==  "M"))]))-(rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$BH ==  "N"))]))
DMPs_MNp001 <- subset(DMPs_MN, DMPs_MN$P.Value<0.001); dim(DMPs_MNp001)
# [1] 394  33

DMPs_CD <- topTable(fit2, num=Inf, coef=4, genelist = ann450kSub)
DMPs_CD$meanDiff <- (rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$BH ==  "C"))]))-(rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$BH ==  "D"))]))
DMPs_CD_sig <- subset(DMPs_CD, DMPs_CD$adj.P.Val <0.05); dim(DMPs_CD_sig)
DMPs_CDp001 <- subset(DMPs_CD, DMPs_CD$P.Value<0.001); dim(DMPs_CDp001)
# DMPs_CDp05 <- subset(DMPs_CD, DMPs_CD$P.Value<0.05); dim(DMPs_CDp05)

GRsetPbmcFunFltCD <- GRsetPbmcFunFlt[,colnames(GRsetPbmcFunFlt)%in%c(row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH == "C"),]), row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH=="D"),]))]


if (require(minfiData)) {
grp <- pData(GRsetPbmcFunFltCD)$BH
M <- getM(GRsetPbmcFunFltCD)
dmpCD <- dmpFinder(M, pheno=grp, type="categorical", shrinkVar = TRUE)
sum(dmpCD$qval < 0.05, na.rm=TRUE)
head(dmpCD)

}
write.csv(DMPs_CNp001, file = "DMPs_BH_CNp001_pbmc.csv")
write.csv(DMPs_DNp001, file = "DMPs_BH_DNp001_pbmc.csv")
write.csv(DMPs_MNp001, file = "DMPs_BH_MNp001_pbmc.csv")
write.csv(DMPs_CDp001, file = "DMPs_BH_CDp001_pbmc.csv")

save(DMPs_CN,DMPs_DN,DMPs_MN,DMPs_CD, file = "DMPs_BH_pbmc.Rda")
```

# DMRs unsing CombP (not used)

```{r}
PbmcDxDMPs_sig_p001 <- subset(PbmcDxDMPs, PbmcDxDMPs$pval<0.001)

df1 <- data.frame(seqnames=PbmcDxDMPs_sig_p001$chr,
  starts=PbmcDxDMPs_sig_p001$pos-1,
  ends=PbmcDxDMPs_sig_p001$pos,
  names=c(rep(".", dim(PbmcDxDMPs_sig_p001)[1])),
  scores=c(rep(".", dim(PbmcDxDMPs_sig_p001)[1])),
  strands=PbmcDxDMPs_sig_p001$strand,
  pvalue = PbmcDxDMPs_sig_p001$pval)

write.table(df1, file="df.bed", quote=F, sep="\t", row.names=F, col.names=F)

```

```{r setup}
# Set working directory outside the chunks
knitr::opts_knit$set(root.dir = normalizePath("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/"))
```
# DMRs # C and D only
```{r}
# targets <- pData(GRsetPbmcFunFltCD)
bh <- factor(pData(GRsetPbmcFunFltCD)$BH)
design <- model.matrix(~0+bh, data=pData(GRsetPbmcFunFltCD))
colnames(design) <- levels(bh)
contMatrix <- makeContrasts(C-D, levels=design)
contMatrix
myAnnotation <- cpg.annotate(object = getM(GRsetPbmcFunFltCD), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "C - D", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : Factor w/ 420257 levels "cg00000029","cg00000108",..: 222425 374552 248925 18571 277135 194037 362374 266425 100898 55258 ...
#  $ stat  : num [1:420257] 0.988 0.626 -0.511 -0.166 0.138 ...
#  $ CHR   : Factor w/ 22 levels "chr1","chr10",..: 1 1 1 1 1 1 1 1 1 1 ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 0.00969 0.004969 -0.003749 -0.000111 0.000572 ...
#  $ indfdr: num [1:420257] 0.99 0.991 0.991 0.997 0.998 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
results.ranges.sort<- data.frame(results.ranges)[order(-data.frame(results.ranges)$no.cpgs),]
write.csv(results.ranges.sort, file= "DMR_bowelHabits_CD_pbmc.csv")

```

```{r}
## NR4A2 DMR
NR4A2SigCpgs <- DMPs_CDp05[DMPs_CDp05$UCSC_RefGene_Name =="NR4A2",]
NR4A2SigCpgsPlusStr <- subset(NR4A2SigCpgs, NR4A2SigCpgs$strand == "+")
NR4A2SigCpgsNegStr <- subset(NR4A2SigCpgs, NR4A2SigCpgs$strand == "-")

NR4A2SigCpgs <- NR4A2SigCpgsNegStr
NR4A2SigCpgs <- NR4A2SigCpgs[order(NR4A2SigCpgs$pos),]
betas_NR4A2dmr <- getBeta(GRsetPbmcFunFltCD)[row.names(getBeta(GRsetPbmcFunFltCD))%in%row.names(NR4A2SigCpgs),]
bh <- as.factor(pData(GRsetPbmcFunFltCD)$BH)
betas_NR4A2dmr <- betas_NR4A2dmr[row.names(NR4A2SigCpgs),]
grObj <- makeGRangesFromDataFrame(data.frame(chr = NR4A2SigCpgs$chr, start = NR4A2SigCpgs$pos, end = NR4A2SigCpgs$pos+1, strand = NR4A2SigCpgs$strand, score = NR4A2SigCpgs$meanDiff))


atrack <- AnnotationTrack(grObj, name = "CpG")
plotTracks(atrack)
chromosome(atrack) <- 'chr2'

gtrack <- GenomeAxisTrack()
plotTracks(list(gtrack, atrack))

# genome : "hg19" 
# gen<-genome(grObj)

gen = "hg19" 
chr <- as.character(unique(seqnames(grObj)))

# Ideogram track
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
plotTracks(list(itrack, gtrack, atrack))
# geneModels <- data.frame(c(rep("chr20",2)), c(58206655,58205944),c(58303445,58299695),c(96790,93751), c("+","+"),c("protein coding","Nonsense mediated decay"),c("ENSG00000118946" , "ENSG00000118946"),c("ENSG00000118946" , "ENSG00000118946") ,c("ENST00000377918","ENST00000484979"),c("PCDH17","PCDH17"))
# colnames(geneModels) <- c( "chromosome",    "start",      "end", "width", "strand", "feature","gene",  "exon",      "transcript",     "symbol")

# grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr, name = "Gene Model")
# chromosome(grtrack) <- 'chr13'
# plotTracks(list(itrack, gtrack, atrack, grtrack))

##data track
 # coords <- c(rep(58205001,145),rep(58205045,145), rep(58209133,145))
mcols(grObj) <- betas_NR4A2dmr
 bhCol <- ifelse(bh=="C","red","green")
dtrack <- DataTrack(grObj,groups = as.vector(bh),type= c("b"),aggregateGroups = TRUE,
           aggregation = "mean")
# 
# plotTracks(dtrack, groups = as.vector(bh), type= "b", aggregateGroups = TRUE,
#            aggregation = "max")
plotTracks(list(itrack, gtrack, atrack, dtrack))
```

```{r}
## HOXA5/6 DMR
HOXA5SigCpgs <- DMPs_CDp05[DMPs_CDp05$UCSC_RefGene_Name =="HOXA5",]
HOXA5SigCpgsPlusStr <- subset(HOXA5SigCpgs, HOXA5SigCpgs$strand == "+")
HOXA5SigCpgsNegStr <- subset(HOXA5SigCpgs, HOXA5SigCpgs$strand == "-")

HOXA5SigCpgs <- HOXA5SigCpgsPlusStr
HOXA5SigCpgs <- HOXA5SigCpgs[order(HOXA5SigCpgs$pos),]
betas_HOXA5dmr <- getBeta(GRsetPbmcFunFltCD)[row.names(getBeta(GRsetPbmcFunFltCD))%in%row.names(HOXA5SigCpgs),]
bh <- as.factor(pData(GRsetPbmcFunFltCD)$BH)
grObj <- makeGRangesFromDataFrame(data.frame(chr = HOXA5SigCpgs$chr, start = HOXA5SigCpgs$pos, end = HOXA5SigCpgs$pos+1, strand = HOXA5SigCpgs$strand, score = HOXA5SigCpgs$meanDiff))


atrack <- AnnotationTrack(grObj, name = "CpG")
plotTracks(atrack)
chromosome(atrack) <- 'chr7'

gtrack <- GenomeAxisTrack()
plotTracks(list(gtrack, atrack))

# genome : "hg19" 
# gen<-genome(grObj)

gen = "hg19" 
chr <- as.character(unique(seqnames(grObj)))

# Ideogram track
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
plotTracks(list(itrack, gtrack, atrack))
# geneModels <- data.frame(c(rep("chr20",2)), c(58206655,58205944),c(58303445,58299695),c(96790,93751), c("+","+"),c("protein coding","Nonsense mediated decay"),c("ENSG00000118946" , "ENSG00000118946"),c("ENSG00000118946" , "ENSG00000118946") ,c("ENST00000377918","ENST00000484979"),c("PCDH17","PCDH17"))
# colnames(geneModels) <- c( "chromosome",    "start",      "end", "width", "strand", "feature","gene",  "exon",      "transcript",     "symbol")

# grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr, name = "Gene Model")
# chromosome(grtrack) <- 'chr13'
# plotTracks(list(itrack, gtrack, atrack, grtrack))

##data track
 # coords <- c(rep(58205001,145),rep(58205045,145), rep(58209133,145))

mcols(grObj) <- betas_HOXA5dmr
dtrack <- DataTrack(grObj,groups = as.vector(bh),type= c("b"),aggregateGroups = TRUE,
           aggregation = "mean")
# 
# plotTracks(dtrack, groups = as.vector(bh), type= "b", aggregateGroups = TRUE,
#            aggregation = "max")
plotTracks(list(itrack, gtrack, atrack, dtrack))
```
# DMRs # C and N only
```{r}
GRsetPbmcFunFltCN <- GRsetPbmcFunFlt[,colnames(GRsetPbmcFunFlt)%in%c(row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH == "C"),]), row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH=="N"),]))]
# targets <- pData(GRsetPbmcFunFltCN)
bh <- factor(pData(GRsetPbmcFunFltCN)$BH)
design <- model.matrix(~0+bh, data=pData(GRsetPbmcFunFltCN))
colnames(design) <- levels(bh)
contMatrix <- makeContrasts(C-N, levels=design)
contMatrix
myAnnotation <- cpg.annotate(object = getM(GRsetPbmcFunFltCN), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "C - N", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : Factor w/ 420257 levels "cg00000029","cg00000108",..: 222425 374552 248925 18571 277135 194037 362374 266425 100898 55258 ...
#  $ stat  : num [1:420257] 0.988 0.626 -0.511 -0.166 0.138 ...
#  $ CHR   : Factor w/ 22 levels "chr1","chr10",..: 1 1 1 1 1 1 1 1 1 1 ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 0.00969 0.004969 -0.003749 -0.000111 0.000572 ...
#  $ indfdr: num [1:420257] 0.99 0.991 0.991 0.997 0.998 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
results.ranges.sort<- data.frame(results.ranges)[order(-data.frame(results.ranges)$no.cpgs),]
write.csv(results.ranges.sort, file= "DMR_bowelHabits_CN_pbmc.csv")
```
# DMRs # D and N only
```{r}
GRsetPbmcFunFltDN <- GRsetPbmcFunFlt[,colnames(GRsetPbmcFunFlt)%in%c(row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH == "D"),]), row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH=="N"),]))]
# targets <- pData(GRsetPbmcFunFltDN)
bh <- factor(pData(GRsetPbmcFunFltDN)$BH)
design <- model.matrix(~0+bh, data=pData(GRsetPbmcFunFltDN))
colnames(design) <- levels(bh)
contMatrix <- makeContrasts(D-N, levels=design)
contMatrix
myAnnotation <- cpg.annotate(object = getM(GRsetPbmcFunFltDN), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "D - N", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : Factor w/ 420257 levels "cg00000029","cg00000108",..: 222425 374552 248925 18571 277135 194037 362374 266425 100898 55258 ...
#  $ stat  : num [1:420257] 0.988 0.626 -0.511 -0.166 0.138 ...
#  $ CHR   : Factor w/ 22 levels "chr1","chr10",..: 1 1 1 1 1 1 1 1 1 1 ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 0.00969 0.004969 -0.003749 -0.000111 0.000572 ...
#  $ indfdr: num [1:420257] 0.99 0.991 0.991 0.997 0.998 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
results.ranges.sort<- data.frame(results.ranges)[order(-data.frame(results.ranges)$no.cpgs),]
write.csv(results.ranges.sort, file= "DMR_bowelHabits_DN_pbmc.csv")
```
# DMRs # M and N only
```{r}
GRsetPbmcFunFltMN <- GRsetPbmcFunFlt[,colnames(GRsetPbmcFunFlt)%in%c(row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH == "M"),]), row.names(pData(GRsetPbmcFunFlt)[which(pData(GRsetPbmcFunFlt)$BH=="N"),]))]
# targets <- pData(GRsetPbmcFunFltMN)
bh <- factor(pData(GRsetPbmcFunFltMN)$BH)
design <- model.matrix(~0+bh, data=pData(GRsetPbmcFunFltMN))
colnames(design) <- levels(bh)
contMatrix <- makeContrasts(M-N, levels=design)
contMatrix
myAnnotation <- cpg.annotate(object = getM(GRsetPbmcFunFltMN), datatype = "array", what = "M", analysis.type = "differential", design = design, contrasts = TRUE, cont.matrix = contMatrix, coef = "M - N", arraytype = "450K")

str(myAnnotation)
# List of 7
#  $ ID    : Factor w/ 420257 levels "cg00000029","cg00000108",..: 222425 374552 248925 18571 277135 194037 362374 266425 100898 55258 ...
#  $ stat  : num [1:420257] 0.988 0.626 -0.511 -0.166 0.138 ...
#  $ CHR   : Factor w/ 22 levels "chr1","chr10",..: 1 1 1 1 1 1 1 1 1 1 ...
#  $ pos   : int [1:420257] 15865 534242 710097 714177 720865 758829 763119 779995 805102 805338 ...
#  $ betafc: num [1:420257] 0.00969 0.004969 -0.003749 -0.000111 0.000572 ...
#  $ indfdr: num [1:420257] 0.99 0.991 0.991 0.997 0.998 ...
#  $ is.sig: logi [1:420257] FALSE FALSE FALSE FALSE FALSE FALSE ...
#  - attr(*, "row.names")= int [1:420257] 1 2 3 4 5 6 7 8 9 10 ...
#  - attr(*, "class")= chr "annot"

DMRs <- dmrcate(myAnnotation, lambda=1000, C=2, pcutoff = 0.001)
head(DMRs$results); dim(DMRs$results)
DMRsInt<- DMRs$results[order(-DMRs$results$no.cpgs),]

# convert the regions to annotated genomic ranges
# data(dmrcatedata)
results.ranges <- extractRanges(DMRs, genome = "hg19")
results.ranges.sort<- data.frame(results.ranges)[order(-data.frame(results.ranges)$no.cpgs),]
write.csv(results.ranges.sort, file= "DMR_bowelHabits_MN_pbmc.csv")
```




```{r setup}
# Set working directory outside the chunks
knitr::opts_knit$set(root.dir = normalizePath("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/"))
```

# Heatmap and boxplot
```{r}
# heatmap
png("temp/functional_norm/methBioconductorWorkflow/pbmc/PbmcDxp001heatmap.png", height = 2000, width = 2000, res = 250)
heatmapMeth(GRsetPbmcFunFlt, row.names(PbmcDxDMPs_sig_p001), pData(GRsetPbmcFunFlt)$Dx, pData(GRsetPbmcFunFlt)$BH, pData(GRsetPbmcFunFlt)$Gender)
dev.off()

# box plots
library(ggplot2)
# plot the top 4 most significantly differentially methylated CpGs 
PbmcDxDMPs_sig_p001 <- PbmcDxDMPs_sig_p001[order(-PbmcDxDMPs_sig_p001$meanDiff),]
dmp <- 1:16
df <- cbind(as.data.frame(t(getBeta(GRsetPbmcFunFlt)[row.names(GRsetPbmcFunFlt)%in%rownames(PbmcDxDMPs_sig_p001)[dmp],])), pData(GRsetPbmcFunFlt)$BH, pData(GRsetPbmcFunFlt)$Dx) 

colnames(df)[-c(dim(df)[2],dim(df)[2]-1)] <- PbmcDxDMPs_sig_p001[dmp,][,26]
colnames(df)[which(duplicated(colnames(df)))] <- paste(colnames(df)[which(duplicated(colnames(df)))],"_1",sep="")
colnames(df)[dim(df)[2]] <- "Dx"
colnames(df)[dim(df)[2]-1] <- "BH"
colnames(df) <- gsub("-","_",colnames(df))

library(gridExtra)   
png("temp/functional_norm/methBioconductorWorkflow/pbmc/Pbmc_Dx001p_dmp1_16.png", height = 3000, width  =3000,res =250)

plots = NULL
for (cpg in colnames(df)[-c(dim(df)[2],dim(df)[2]-1)]) {
        plots[[cpg]] = ggplot(df, aes_string("Dx", cpg)) + geom_boxplot(outlier.shape = NA) + geom_jitter(aes(color = df$BH))
        }

grid.arrange(plots[[1]],plots[[2]],plots[[3]],plots[[4]],
             plots[[5]],plots[[6]],plots[[7]],plots[[8]],
             plots[[9]],plots[[10]],plots[[11]],plots[[12]],
             plots[[13]],plots[[14]],plots[[15]],plots[[16]],
            ncol = 4)
dev.off()
```


# Dx Gender and interaction assocaiated DMPs
```{r}
# # Dx Gender interaction
# pData(GRsetPbmcFunFlt)$ibsGenInt <- interaction(pData(GRsetPbmcFunFlt)$Dx,pData(GRsetPbmcFunFlt)$Gender)
# pData(GRsetPbmcFunFlt)$ibsGenInt <- gsub("\\.","_", pData(GRsetPbmcFunFlt)$ibsGenInt)
# dmpsIntPbmc <- DMPsInt(GRsetPbmcFunFlt,factor(pData(GRsetPbmcFunFlt)$ibsGenInt), ann450kSub)

mVals_p <- getM(GRsetPbmcFunFlt)
head(mVals_p[,1:5])
betaVals_p <- getBeta(GRsetPbmcFunFlt)
head(betaVals_p[,1:5])

# factor of interest
Dx <- factor(pData(GRsetPbmcFunFlt)$Dx)
Gender <- factor(pData(GRsetPbmcFunFlt)$Gender)

# variable to correct for- coupld potenitally correct for age, but leave it out

# use the above to create a design matrix
design <- model.matrix(~ 0 + Dx:Gender, data=pData(GRsetPbmcFunFlt))
colnames(design) <- c("HC_F", "IBS_F", "HC_M", "IBS_M")

# fit the linear model 
fit <- lmFit(mVals_p, design)

# create a contrast matrix for specific comparisons
contMatrix <- makeContrasts(IBS_M-HC_M, IBS_F-HC_F,IBS_M-IBS_F, HC_M-HC_F,   levels=design)
contMatrix

# fit the contrasts
fit2 <- contrasts.fit(fit, contMatrix)
fit2 <- eBayes(fit2)

# look at the numbers of DM CpGs at FDR < 0.05
summary(decideTests(fit2))
# 
#       IBS_M - HC_M IBS_F - HC_F IBS_M - IBS_F HC_M - HC_F
# -1            0            2         17882         416
# 0        420242       420255        379977      419804
# 1            15            0         22398          37

# get the table of results for the first contrast (naive - rTreg)
ann450kSub <- hm450annotations[match(rownames(mVals_p),hm450annotations$Name),
                      c(1:4,12:19,24:ncol(hm450annotations))]
DMPs_IBS_M_HC_M <-  topTable(fit2, num=Inf, coef=1, genelist=ann450kSub)
DMPs_IBS_M_HC_M$meanDiff <- rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$ibsGenInt ==  "IBS_M"))])-rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$ibsGenInt ==  "HC_M"))])
DMPs_IBS_M_HC_MSig <- subset(DMPs_IBS_M_HC_M, DMPs_IBS_M_HC_M$adj.P.Val<0.05); dim(DMPs_IBS_M_HC_MSig)

DMPs_IBS_F_HC_F <-  topTable(fit2, num=Inf, coef=2, genelist=ann450kSub)
DMPs_IBS_F_HC_F$meanDiff <- rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$ibsGenInt ==  "IBS_F"))])-rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$ibsGenInt ==  "HC_F"))])
DMPs_IBS_F_HC_FSig <- subset(DMPs_IBS_F_HC_F, DMPs_IBS_F_HC_F$adj.P.Val<0.05); dim(DMPs_IBS_F_HC_FSig)

DMPs_IBS_M_IBS_F <-  topTable(fit2, num=Inf, coef=3, genelist=ann450kSub)
DMPs_IBS_M_IBS_F$meanDiff <- rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$ibsGenInt ==  "IBS_M"))])-rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$ibsGenInt ==  "IBS_F"))])

DMPs_HC_M_HC_F <- topTable(fit2, num=Inf, coef=4, genelist=ann450kSub)
DMPs_HC_M_HC_F$meanDiff <- rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$ibsGenInt ==  "HC_M"))])-rowMeans(betaVals_p[,colnames(betaVals_p)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$ibsGenInt ==  "HC_F"))])

# remove HC Gender probes from IBS
DMPs_IBS_M_IBS_F_sig <- subset(DMPs_IBS_M_IBS_F, DMPs_IBS_M_IBS_F$adj.P.Val<0.05); dim(DMPs_IBS_M_IBS_F_sig)
# [1] 40280    33

DMPs_HC_M_HC_F_sig <- subset(DMPs_HC_M_HC_F, DMPs_HC_M_HC_F$adj.P.Val<0.05); dim(DMPs_HC_M_HC_F_sig)
# [1] 453  33

DMPs_HC_M_HC_F_sig <- subset(DMPs_HC_M_HC_F, DMPs_HC_M_HC_F$P.Value<0.05); dim(DMPs_HC_M_HC_F_sig)
# [1] 30448    33


DMPs_IBS_M_IBS_F_sig1 <- DMPs_IBS_M_IBS_F_sig[!row.names(DMPs_IBS_M_IBS_F_sig)%in%row.names(DMPs_HC_M_HC_F_sig),]; dim(DMPs_IBS_M_IBS_F_sig1)

# [1] 37363  32
```

```{r}

# PbmcGenderDMPs <- DMPs(GRsetPbmcFunFlt,factor(pData(GRsetPbmcFunFlt)$Gender), ann450kSub)
# PbmcGenderDMPs_sig <- subset(PbmcGenderDMPs, PbmcGenderDMPs$adj.P.Val<0.05); dim(PbmcGenderDMPs_sig)

grp <- pData(GRsetPbmcFunFlt)$Gender
M <- getM(GRsetPbmcFunFlt)
B <- getBeta(GRsetPbmcFunFlt)
PbmcSexDMPs <- dmpFinder(M, pheno=grp, type="categorical", shrinkVar = TRUE)
PbmcSexDMPs$Bdiff <- (rowMeans(B[row.names(PbmcSexDMPs),colnames(B)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$Gender == "F"))]))-(rowMeans(B[row.names(PbmcSexDMPs),colnames(B)%in%row.names(subset(pData(GRsetPbmcFunFlt), pData(GRsetPbmcFunFlt)$Gender == "M"))]))
sum(PbmcSexDMPs$qval < 0.05, na.rm=TRUE)
head(PbmcSexDMPs)
ann450kSub <- ann450kSub[row.names(ann450kSub)%in%row.names(PbmcSexDMPs),]
ann450kSub <- ann450kSub[row.names(PbmcSexDMPs),]
PbmcSexDMPs <- cbind(PbmcSexDMPs,ann450kSub)
PbmcSexDMPsSig <- subset(PbmcSexDMPs,PbmcSexDMPs$qval < 0.05 )
```

```{r}
DMPs_IBS_M_IBS_F_sig2 <- DMPs_IBS_M_IBS_F_sig1[!row.names(DMPs_IBS_M_IBS_F_sig1)%in%row.names(PbmcSexDMPsSig),]; dim(DMPs_IBS_M_IBS_F_sig2)
# [1] 27128    33

DMPs_IBS_M_IBS_F_sig3 <- subset(DMPs_IBS_M_IBS_F_sig2, DMPs_IBS_M_IBS_F_sig2$adj.P.Val<0.01);dim(DMPs_IBS_M_IBS_F_sig3)
# [1] 4072   32

# DMPs_IBS_M_IBS_F_sig3 <- subset(DMPs_IBS_M_IBS_F_sig2, DMPs_IBS_M_IBS_F_sig2$meanDiff>=0.02|DMPs_IBS_M_IBS_F_sig2$meanDiff< -0.02); dim(DMPs_IBS_M_IBS_F_sig3)
# 
# save(DMPs_IBS_M_IBS_F_sig2, file = "C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/pbmc/DMPs_IBS_M_IBS_F_sig2.rda")
# 
# write.csv(DMPs_IBS_M_IBS_F_sig3, file = "C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/pbmc/DMPs_IBS_M_IBS_F_sig2_28.csv")

```

```{r}
# ## same results from making an interaction variable.
# 
# dmpsIntPbmc <- read.csv("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/temp/functional_norm/methBioconductorWorkflow/pbmc/dmpsIntPbmc.csv", row.names = 4)
# dmpsIntPbmc <- as.data.frame(dmpsIntPbmc)
# 
# ibs_m_f_sig <- subset(dmpsIntPbmc, dmpsIntPbmc$dmps_IBS_M__IBS_F.adj.P.Val<0.05); dim(ibs_m_f_sig)
# # [1] 40280    33
# hc_m_f_sig <- subset(dmpsIntPbmc[[4]], dmpsIntPbmc[[4]]$adj.P.Val<0.05); dim(hc_m_f_sig)
# # # [1] 453  33 
# ibs_m_f_sig1 <- ibs_m_f_sig[!row.names(ibs_m_f_sig)%in%row.names(hc_m_f_sig),]; dim(ibs_m_f_sig1)
# # # [1] 39990    33
# 
# ## very few gender differences in HC compared to IBS.use p value instead to remove as many gender assocaited probes as possible.
# hc_m_f_sig <- subset(dmpsIntPbmc[[4]], dmpsIntPbmc[[4]]$P.Value<0.05); dim(hc_m_f_sig)
# # [1] 30448    33
# ibs_m_f_sig1 <- ibs_m_f_sig[!row.names(ibs_m_f_sig)%in%row.names(hc_m_f_sig),]; dim(ibs_m_f_sig1)
# # [1] 37363    33
# 
# # IBS HC differences within 1. Women 2. men
# f_ibs_hc_sig <- subset(dmpsIntPbmc[[1]], dmpsIntPbmc[[1]]$adj.P.Val<0.05); dim(f_ibs_hc_sig)
# # [1]  2 33
# m_ibs_hc_sig<- subset(dmpsIntPbmc[[2]], dmpsIntPbmc[[2]]$adj.P.Val<0.05); dim(m_ibs_hc_sig)
# # [1] 15 33
# 
# ## try to remove general gender differences
# ibs_m_f_sig2 <- ibs_m_f_sig[!row.names(ibs_m_f_sig)%in%row.names(PbmcGenderDMPs_sig),]; dim(ibs_m_f_sig2)
# # [1] 31759    33 
# ibs_m_f_sig3 <- ibs_m_f_sig2[!row.names(ibs_m_f_sig2)%in%row.names(hc_m_f_sig),]; dim(ibs_m_f_sig3)
# # [1] 30821    33 ### final
# 
# ibs_m_f_sig3fdr01 <- subset(ibs_m_f_sig3, ibs_m_f_sig3$adj.P.Val<0.01); dim(ibs_m_f_sig3fdr01)
# # [1] 5788   33
# 
# # ibs_m_f_sig3fdr01md <- subset(ibs_m_f_sig3fdr01, ibs_m_f_sig3fdr01$meanDiff >0.3 | ibs_m_f_sig3fdr01$meanDiff < -0.3); dim(ibs_m_f_sig3fdr01md )

```

```{r}
# Set working directory outside the chunks
knitr::opts_knit$set(root.dir = normalizePath("C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/"))
```

```{r}

write.table(PbmcDxDMPs, file="PbmcDxDMPs.csv", sep=",", row.names=FALSE)

write.table(PbmcDxDMPs_sig_p001, file="PbmcDxDMPs_sig_p001.csv", sep=",", row.names=FALSE)
```

```{r}
#################### save the results

write.table(PbmcSexDMPsSig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/PbmcGenderDMPs.csv", sep=",", row.names=FALSE)

save(PbmcSexDMPs, file = "C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/PbmcSexDMPs.rda")

save(fit2, file = "C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/fitted_gender_dx_int_pbmc.rda")

write.table(DMPs_IBS_F_HC_FSig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/Pbmc_DMPs_IBS_F_HC_FSig.csv", sep=",", row.names=FALSE)

write.table(DMPs_IBS_M_HC_MSig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/Pbmc_DMPs_IBS_M_HC_MSig.csv", sep=",", row.names=FALSE)

write.table(DMPs_IBS_M_IBS_F_sig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/DMPs_IBS_M_IBS_F_sig.csv", sep=",", row.names=FALSE)

 write.table(DMPs_HC_M_HC_F_sig, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/DMPs_HC_M_HC_F_sig.csv", sep=",", row.names=FALSE)

write.table(DMPs_IBS_M_IBS_F_sig2, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/DMPs_IBS_M_IBS_F_sig2.csv", sep=",", row.names=FALSE)

write.table(DMPs_IBS_M_IBS_F_sig3, file="C:/Users/swapnajoshi-admin/Documents/GIT_workspace/IBSdnaMethylationProject/temp1/res3_functional_norm/methBioconductorWorkflow/pbmc/DMPs_IBS_M_IBS_F_sig3.csv", sep=",", row.names=FALSE)

# save(dmpsIntPbmc , file = "temp/functional_norm/methBioconductorWorkflow/pbmc/Pbmc_dmpsIntPbmc.rda")
```
# Table 1: Demographics of patients compared to controls
```{r}
t_p <- function (modelobject) {
     p <- modelobject$p.value
    return(p)
}
metDat_p<- read.csv("C:/Users/swapnajoshi-admin/Box Sync/DNA_methylation_R21_Analysis/data/rawdata/Metadata_pbmc_methylation.csv", row.names = 1)
load("temp/functional_norm/methBioconductorWorkflow/GRsetPbmcColFunFlt.Rda")
metDat_p <- metDat_p[row.names(metDat_p)%in%substr(colnames(GRsetPbmcFunFlt),1,5),-c(22,23,24,26,28)]; dim(metDat_p)
met.hc <- subset(metDat_p, metDat_p$Group==1); dim(met.hc)
met.ibs <- subset(metDat_p, metDat_p$Group==2); dim(met.ibs)
table1 <- matrix(NA, nrow = 23, ncol = 3)

table1[c(4:23),1] <- round(apply(met.ibs[,c(4:23)],2,mean, na.rm = TRUE),2)
table1[c(4:23),2] <- round(apply(met.hc[,c(4:23)],2,mean, na.rm = TRUE),2)
table1[3,1] <- round(table(met.ibs$Sex)[2]/(table(met.ibs$Sex)[2]+table(met.ibs$Sex)[1])*100,2)
table1[3,2] <-round(table(met.hc$Sex)[2]/(table(met.hc$Sex)[2]+table(met.hc$Sex)[1])*100,2)

row.names(table1)[c(4:23)] <- colnames(met.ibs)[c(4:23)]
row.names(table1)[3]<-"Female %"
fisher.test(table(metDat_p$Group,metDat_p$Sex))

metDat_p$Group <- as.factor(metDat_p$Group)
comb1 <- expand.grid(names(metDat_p[,c(4,5,8,9,10, 11,12,13,14,15,16,17,18,19,20,22,23)]),"Group")

models <- lapply(paste(comb1[,1],comb1[,2], sep = "~"), formula)
res.models <- lapply(models, FUN = function(x) {t.test(formula = x, data = metDat_p)})
# names(res.models) <- paste(comb1[,1],comb1[,2], sep = "~")
table1[c(4,5,8,9,10, 11,12,13,14,15,16,17,18,19,20,22,23),3] <- round(t(data.frame(lapply(res.models, FUN = t_p))),6)
table1[3,3] <- round(fisher.test(table(metDat_p$Group,metDat_p$Sex))$p.value,3)
colnames(table1) <- c("IBS (N=105)","Healthy Controls (N=36)", "t.test p-value")
table2 <- matrix(as.numeric(as.character(table1)), nrow = 23, ncol = 3)
row.names(table2) <- row.names(table1)
colnames(table2) <- colnames(table1)
table2[c(2,6,7,21),c(2,3)] <- "-"
table2[2,1] <- paste(round(table(met.ibs$BH_Blood_Collection)[1]/(table(met.ibs$BH_Blood_Collection)[2]+table(met.ibs$BH_Blood_Collection)[1] + table(met.ibs$BH_Blood_Collection)[3])*100,2),round(table(met.ibs$BH_Blood_Collection)[2]/(table(met.ibs$BH_Blood_Collection)[2]+table(met.ibs$BH_Blood_Collection)[1] + table(met.ibs$BH_Blood_Collection)[3])*100,2),round(table(met.ibs$BH_Blood_Collection)[3]/(table(met.ibs$BH_Blood_Collection)[2]+table(met.ibs$BH_Blood_Collection)[1] + table(met.ibs$BH_Blood_Collection)[3])*100,2), sep = ",")
row.names(table2)[2]<-"Bowe Habits, IBS-C, IBS-D, IBS-M (%)"
library(knitr)
kable(table2,align=c(rep('c', 23)), caption = "Table1: Clinical Characteristics of IBS Patients and HCs")

```